// @telegram-apps/sdk-react/dist/index.js
import { useSyncExternalStore as Aa } from "react";
function of(e, t) {
  return Aa(
    (n) => e.sub(n),
    e,
    t || e
  );
}
// @__NO_SIDE_EFFECTS__
function Fr(e) {
  return {
    lang: (e == null ? void 0 : e.lang) ?? void 0,
    message: e == null ? void 0 : e.message,
    abortEarly: (e == null ? void 0 : e.abortEarly) ?? void 0,
    abortPipeEarly: (e == null ? void 0 : e.abortPipeEarly) ?? void 0
  };
}
// @__NO_SIDE_EFFECTS__
function Ca(e, t) {
  var n;
  return (n = void 0) == null ? void 0 : n.get(t);
}
// @__NO_SIDE_EFFECTS__
function xa(e) {
  var t, n;
  const r = typeof e;
  return r === "string" ? `"${e}"` : r === "number" || r === "bigint" || r === "boolean" ? `${e}` : r === "object" || r === "function" ? (e && ((n = (t = Object.getPrototypeOf(e)) == null ? void 0 : t.constructor) == null ? void 0 : n.name)) ?? "null" : r;
}
function ge(e, t, n, r, s) {
  const o = s && "input" in s ? s.input : n.value, a = (s == null ? void 0 : s.expected) ?? e.expects ?? null, i = (s == null ? void 0 : s.received) ?? /* @__PURE__ */ xa(o), u = {
    kind: e.kind,
    type: e.type,
    input: o,
    expected: a,
    received: i,
    message: `Invalid ${t}: ${a ? `Expected ${a} but r` : "R"}eceived ${i}`,
    requirement: e.requirement,
    path: s == null ? void 0 : s.path,
    issues: s == null ? void 0 : s.issues,
    lang: r.lang,
    abortEarly: r.abortEarly,
    abortPipeEarly: r.abortPipeEarly
  }, c = e.kind === "schema", l = (s == null ? void 0 : s.message) ?? e.message ?? /* @__PURE__ */ Ca(e.reference, u.lang) ?? (c ? (u.lang, void 0) : null) ?? r.message ?? (u.lang, void 0);
  l && (u.message = typeof l == "function" ? (
    // @ts-expect-error
    l(u)
  ) : l), c && (n.typed = !1), n.issues ? n.issues.push(u) : n.issues = [u];
}
// @__NO_SIDE_EFFECTS__
function K(e) {
  return {
    version: 1,
    vendor: "valibot",
    validate(t) {
      return e["~run"]({ value: t }, /* @__PURE__ */ Fr());
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Pa(e, t) {
  return Object.hasOwn(e, t) && t !== "__proto__" && t !== "prototype" && t !== "constructor";
}
var Ba = class extends Error {
  /**
   * Creates a Valibot error with useful information.
   *
   * @param issues The error issues.
   */
  constructor(e) {
    super(e[0].message), this.name = "ValiError", this.issues = e;
  }
};
// @__NO_SIDE_EFFECTS__
function Xt(e, t, n) {
  return typeof e.default == "function" ? (
    // @ts-expect-error
    e.default(t, n)
  ) : (
    // @ts-expect-error
    e.default
  );
}
// @__NO_SIDE_EFFECTS__
function Xe(e, t) {
  return !e["~run"]({ value: t }, { abortEarly: !0 }).issues;
}
// @__NO_SIDE_EFFECTS__
function Wr() {
  return {
    kind: "schema",
    type: "any",
    reference: Wr,
    expects: "any",
    async: !1,
    get "~standard"() {
      return /* @__PURE__ */ K(this);
    },
    "~run"(e) {
      return e.typed = !0, e;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Mt(e) {
  return {
    kind: "schema",
    type: "boolean",
    reference: Mt,
    expects: "boolean",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ K(this);
    },
    "~run"(t, n) {
      return typeof t.value == "boolean" ? t.typed = !0 : ge(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function en(e) {
  return {
    kind: "schema",
    type: "function",
    reference: en,
    expects: "Function",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ K(this);
    },
    "~run"(t, n) {
      return typeof t.value == "function" ? t.typed = !0 : ge(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function N(e, t) {
  return {
    kind: "schema",
    type: "loose_object",
    reference: N,
    expects: "Object",
    async: !1,
    entries: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ K(this);
    },
    "~run"(n, r) {
      var s;
      const o = n.value;
      if (o && typeof o == "object") {
        n.typed = !0, n.value = {};
        for (const a in this.entries) {
          const i = this.entries[a];
          if (a in o || (i.type === "exact_optional" || i.type === "optional" || i.type === "nullish") && // @ts-expect-error
          i.default !== void 0) {
            const u = a in o ? (
              // @ts-expect-error
              o[a]
            ) : /* @__PURE__ */ Xt(i), c = i["~run"]({ value: u }, r);
            if (c.issues) {
              const l = {
                type: "object",
                origin: "value",
                input: o,
                key: a,
                value: u
              };
              for (const p of c.issues)
                p.path ? p.path.unshift(l) : p.path = [l], (s = n.issues) == null || s.push(p);
              if (n.issues || (n.issues = c.issues), r.abortEarly) {
                n.typed = !1;
                break;
              }
            }
            c.typed || (n.typed = !1), n.value[a] = c.value;
          } else if (i.type !== "exact_optional" && i.type !== "optional" && i.type !== "nullish" && (ge(this, "key", n, r, {
            input: void 0,
            expected: `"${a}"`,
            path: [
              {
                type: "object",
                origin: "key",
                input: o,
                key: a,
                // @ts-expect-error
                value: o[a]
              }
            ]
          }), r.abortEarly))
            break;
        }
        if (!n.issues || !r.abortEarly)
          for (const a in o)
            /* @__PURE__ */ Pa(o, a) && !(a in this.entries) && (n.value[a] = o[a]);
      } else
        ge(this, "type", n, r);
      return n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Be(e, t) {
  return {
    kind: "schema",
    type: "nullish",
    reference: Be,
    expects: `(${e.expects} | null | undefined)`,
    async: !1,
    wrapped: e,
    default: t,
    get "~standard"() {
      return /* @__PURE__ */ K(this);
    },
    "~run"(n, r) {
      return (n.value === null || n.value === void 0) && (this.default !== void 0 && (n.value = /* @__PURE__ */ Xt(this, n, r)), n.value === null || n.value === void 0) ? (n.typed = !0, n) : this.wrapped["~run"](n, r);
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Tt(e) {
  return {
    kind: "schema",
    type: "number",
    reference: Tt,
    expects: "number",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ K(this);
    },
    "~run"(t, n) {
      return typeof t.value == "number" && !isNaN(t.value) ? t.typed = !0 : ge(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function $t(e, t) {
  return {
    kind: "schema",
    type: "optional",
    reference: $t,
    expects: `(${e.expects} | undefined)`,
    async: !1,
    wrapped: e,
    default: t,
    get "~standard"() {
      return /* @__PURE__ */ K(this);
    },
    "~run"(n, r) {
      return n.value === void 0 && (this.default !== void 0 && (n.value = /* @__PURE__ */ Xt(this, n, r)), n.value === void 0) ? (n.typed = !0, n) : this.wrapped["~run"](n, r);
    }
  };
}
// @__NO_SIDE_EFFECTS__
function ne(e) {
  return {
    kind: "schema",
    type: "string",
    reference: ne,
    expects: "string",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ K(this);
    },
    "~run"(t, n) {
      return typeof t.value == "string" ? t.typed = !0 : ge(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Nr() {
  return {
    kind: "schema",
    type: "unknown",
    reference: Nr,
    expects: "unknown",
    async: !1,
    get "~standard"() {
      return /* @__PURE__ */ K(this);
    },
    "~run"(e) {
      return e.typed = !0, e;
    }
  };
}
function It(e, t, n) {
  const r = e["~run"]({ value: t }, /* @__PURE__ */ Fr(n));
  if (r.issues)
    throw new Ba(r.issues);
  return r.value;
}
// @__NO_SIDE_EFFECTS__
function Hr(...e) {
  return {
    ...e[0],
    pipe: e,
    get "~standard"() {
      return /* @__PURE__ */ K(this);
    },
    "~run"(t, n) {
      for (const r of e)
        if (r.kind !== "metadata") {
          if (t.issues && (r.kind === "schema" || r.kind === "transformation")) {
            t.typed = !1;
            break;
          }
          (!t.issues || !n.abortEarly && !n.abortPipeEarly) && (t = r["~run"](t, n));
        }
      return t;
    }
  };
}
function Ur(e) {
  return /* @__PURE__ */ Xe(
    /* @__PURE__ */ N({ TelegramWebviewProxy: /* @__PURE__ */ N({ postEvent: /* @__PURE__ */ en() }) }),
    e
  );
}
function zr() {
  try {
    return window.self !== window.top;
  } catch {
    return !0;
  }
}
var ja = Object.defineProperty, qa = (e, t, n) => t in e ? ja(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n, sr = (e, t, n) => qa(e, typeof t != "symbol" ? t + "" : t, n);
function Ma(e) {
  return (t) => t instanceof e;
}
function Gr(e, t) {
  t || (t = []);
  class n extends Error {
    constructor(...s) {
      const o = typeof t == "function" ? t(...s) : typeof t == "string" ? [t] : t || [];
      super(...o), this.name = e;
    }
  }
  return Object.defineProperty(n, "name", { value: e }), [n, Ma(n)];
}
const [Ta, af] = Gr("CancelledError", "Promise was canceled"), [$a, uf] = Gr(
  "TimeoutError",
  (e, t) => [`Timeout reached: ${e}ms`, { cause: t }]
), Qr = Symbol("Resolved");
function or(e) {
  return Array.isArray(e) && e[0] === Qr;
}
function Ia(e) {
  return [Qr, e];
}
function ar(e, t) {
  return e.reject = t.reject, e.abort = t.abort, e;
}
let Jr = class Ot extends Promise {
  constructor(t, n) {
    let r, s;
    super((o, a) => {
      let i, u;
      typeof t == "function" ? (i = t, u = n) : u = t;
      const c = [], l = (d) => (...I) => {
        const ee = d(...I);
        return c.forEach((St) => St()), ee;
      }, p = new AbortController(), { signal: m } = p;
      s = (d) => {
        !m.aborted && p.abort(d);
      };
      const f = () => m.reason, x = (d) => {
        const I = () => {
          d(f());
        };
        m.addEventListener("abort", I, !0);
        const ee = () => {
          m.removeEventListener("abort", I, !0);
        };
        return c.push(ee), ee;
      }, be = l((d) => {
        o(d), s(Ia(d));
      });
      r = l((d) => {
        a(d), s(d);
      }), u || (u = {});
      const { abortSignal: P, rejectOnAbort: te = !0 } = u;
      if (P)
        if (P.aborted) {
          const { reason: d } = P;
          if (te)
            return r(d);
          s(d);
        } else {
          const d = () => {
            s(P.reason);
          };
          P.addEventListener("abort", d), c.push(() => {
            P.removeEventListener("abort", d);
          });
        }
      te && x(a);
      const { timeout: S } = u;
      if (S) {
        const d = setTimeout(() => {
          s(new $a(S));
        }, S);
        c.push(() => {
          clearTimeout(d);
        });
      }
      const k = () => m.aborted, $ = () => or(f()), V = () => {
        const d = f();
        return or(d) ? d[1] : void 0;
      };
      try {
        const d = i && i(be, r, {
          abortReason: f,
          abortSignal: m,
          isAborted: k,
          isResolved: $,
          onAborted: x,
          onResolved: (I) => x(() => {
            $() && I(V());
          }),
          resolved: V,
          throwIfAborted() {
            if (k())
              throw f();
          }
        });
        d instanceof Promise && d.catch(r);
      } catch (d) {
        r(d);
      }
    }), sr(this, "abort"), sr(this, "reject"), this.abort = s, this.reject = r;
  }
  /**
   * Creates a new AbortablePromise instance using an executor, resolving the promise when a result
   * was returned.
   * @param fn - function returning promise result.
   * @param options - additional options.
   */
  static fn(t, n) {
    return new Ot(async (r, s, o) => {
      try {
        r(await t(o));
      } catch (a) {
        s(a);
      }
    }, n);
  }
  static resolve(t) {
    return this.fn(() => t);
  }
  /**
   * @see Promise.reject
   */
  static reject(t) {
    return new Ot((n, r) => {
      r(t);
    });
  }
  /**
   * Aborts the promise with the cancel error.
   */
  cancel() {
    this.abort(new Ta());
  }
  /**
   * @see Promise.catch
   */
  catch(t) {
    return this.then(void 0, t);
  }
  /**
   * @see Promise.finally
   */
  finally(t) {
    return ar(super.finally(t), this);
  }
  /**
   * @see Promise.then
   */
  then(t, n) {
    return ar(super.then(t, n), this);
  }
};
function Oa(e) {
  return e.replace(/_[a-z]/g, (t) => t[1].toUpperCase());
}
function La(e) {
  return Object.entries(e).reduce((t, [n, r]) => (t[Oa(n)] = r, t), {});
}
function Lt(e) {
  const t = La(e);
  for (const n in t) {
    const r = t[n];
    r && typeof r == "object" && !(r instanceof Date) && (t[n] = Array.isArray(r) ? r.map(Lt) : Lt(r));
  }
  return t;
}
function Kr(e) {
  return `tapps/${e}`;
}
function Zr(e, t) {
  sessionStorage.setItem(Kr(e), JSON.stringify(t));
}
function Da(e) {
  const t = sessionStorage.getItem(Kr(e));
  try {
    return t ? JSON.parse(t) : void 0;
  } catch {
  }
}
function Ra(...e) {
  const t = e.flat(1);
  return [
    t.push.bind(t),
    () => {
      t.forEach((n) => {
        n();
      });
    }
  ];
}
// @__NO_SIDE_EFFECTS__
function Va(e, t) {
  t || (t = {});
  const {
    textColor: n,
    bgColor: r,
    shouldLog: s
  } = t, o = s === void 0 ? !0 : s, a = typeof o == "boolean" ? () => o : o;
  function i(u, c, ...l) {
    if (c || a()) {
      const p = "font-weight:bold;padding:0 5px;border-radius:5px";
      console[u](
        `%c${Intl.DateTimeFormat("en-GB", {
          hour: "2-digit",
          minute: "2-digit",
          second: "2-digit",
          fractionalSecondDigits: 3,
          timeZone: "UTC"
        }).format(/* @__PURE__ */ new Date())}%c / %c${e}`,
        `${p};background-color: lightblue;color:black`,
        "",
        `${p};${n ? `color:${n};` : ""}${r ? `background-color:${r}` : ""}`,
        ...l
      );
    }
  }
  return [i.bind(void 0, "log"), i.bind(void 0, "error")];
}
// @__NO_SIDE_EFFECTS__
function Yr(e) {
  return {
    lang: (e == null ? void 0 : e.lang) ?? void 0,
    message: e == null ? void 0 : e.message,
    abortEarly: (e == null ? void 0 : e.abortEarly) ?? void 0,
    abortPipeEarly: (e == null ? void 0 : e.abortPipeEarly) ?? void 0
  };
}
// @__NO_SIDE_EFFECTS__
function Fa(e, t) {
  var n;
  return (n = void 0) == null ? void 0 : n.get(t);
}
// @__NO_SIDE_EFFECTS__
function Wa(e) {
  var t, n;
  const r = typeof e;
  return r === "string" ? `"${e}"` : r === "number" || r === "bigint" || r === "boolean" ? `${e}` : r === "object" || r === "function" ? (e && ((n = (t = Object.getPrototypeOf(e)) == null ? void 0 : t.constructor) == null ? void 0 : n.name)) ?? "null" : r;
}
function O(e, t, n, r, s) {
  const o = s && "input" in s ? s.input : n.value, a = (s == null ? void 0 : s.expected) ?? e.expects ?? null, i = (s == null ? void 0 : s.received) ?? /* @__PURE__ */ Wa(o), u = {
    kind: e.kind,
    type: e.type,
    input: o,
    expected: a,
    received: i,
    message: `Invalid ${t}: ${a ? `Expected ${a} but r` : "R"}eceived ${i}`,
    requirement: e.requirement,
    path: s == null ? void 0 : s.path,
    issues: s == null ? void 0 : s.issues,
    lang: r.lang,
    abortEarly: r.abortEarly,
    abortPipeEarly: r.abortPipeEarly
  }, c = e.kind === "schema", l = (s == null ? void 0 : s.message) ?? e.message ?? /* @__PURE__ */ Fa(e.reference, u.lang) ?? (c ? (u.lang, void 0) : null) ?? r.message ?? (u.lang, void 0);
  l && (u.message = typeof l == "function" ? (
    // @ts-expect-error
    l(u)
  ) : l), c && (n.typed = !1), n.issues ? n.issues.push(u) : n.issues = [u];
}
// @__NO_SIDE_EFFECTS__
function F(e) {
  return {
    version: 1,
    vendor: "valibot",
    validate(t) {
      return e["~run"]({ value: t }, /* @__PURE__ */ Yr());
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Xr(e, t) {
  return Object.hasOwn(e, t) && t !== "__proto__" && t !== "prototype" && t !== "constructor";
}
// @__NO_SIDE_EFFECTS__
function Na(e, t) {
  const n = [...new Set(e)];
  return n.length > 1 ? `(${n.join(` ${t} `)})` : n[0] ?? "never";
}
var Ha = class extends Error {
  /**
   * Creates a Valibot error with useful information.
   *
   * @param issues The error issues.
   */
  constructor(e) {
    super(e[0].message), this.name = "ValiError", this.issues = e;
  }
};
// @__NO_SIDE_EFFECTS__
function es(e, t) {
  return {
    kind: "validation",
    type: "check",
    reference: es,
    async: !1,
    expects: null,
    requirement: e,
    message: t,
    "~run"(n, r) {
      return n.typed && !this.requirement(n.value) && O(this, "input", n, r), n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function ts(e) {
  return {
    kind: "validation",
    type: "integer",
    reference: ts,
    async: !1,
    expects: null,
    requirement: Number.isInteger,
    message: e,
    "~run"(t, n) {
      return t.typed && !this.requirement(t.value) && O(this, "integer", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function se(e) {
  return {
    kind: "transformation",
    type: "transform",
    reference: se,
    async: !1,
    operation: e,
    "~run"(t) {
      return t.value = this.operation(t.value), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function ns(e, t, n) {
  return typeof e.default == "function" ? (
    // @ts-expect-error
    e.default(t, n)
  ) : (
    // @ts-expect-error
    e.default
  );
}
// @__NO_SIDE_EFFECTS__
function Ua(e, t) {
  return !e["~run"]({ value: t }, { abortEarly: !0 }).issues;
}
// @__NO_SIDE_EFFECTS__
function je(e) {
  return {
    kind: "schema",
    type: "boolean",
    reference: je,
    expects: "boolean",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(t, n) {
      return typeof t.value == "boolean" ? t.typed = !0 : O(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function rs(e) {
  return {
    kind: "schema",
    type: "date",
    reference: rs,
    expects: "Date",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(t, n) {
      return t.value instanceof Date ? isNaN(t.value) ? O(this, "type", t, n, {
        received: '"Invalid Date"'
      }) : t.typed = !0 : O(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function ss(e, t) {
  return {
    kind: "schema",
    type: "instance",
    reference: ss,
    expects: e.name,
    async: !1,
    class: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(n, r) {
      return n.value instanceof this.class ? n.typed = !0 : O(this, "type", n, r), n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function tn(e) {
  return {
    kind: "schema",
    type: "lazy",
    reference: tn,
    expects: "unknown",
    async: !1,
    getter: e,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(t, n) {
      return this.getter(t.value)["~run"](t, n);
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Se(e, t) {
  return {
    kind: "schema",
    type: "loose_object",
    reference: Se,
    expects: "Object",
    async: !1,
    entries: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(n, r) {
      var s;
      const o = n.value;
      if (o && typeof o == "object") {
        n.typed = !0, n.value = {};
        for (const a in this.entries) {
          const i = this.entries[a];
          if (a in o || (i.type === "exact_optional" || i.type === "optional" || i.type === "nullish") && // @ts-expect-error
          i.default !== void 0) {
            const u = a in o ? (
              // @ts-expect-error
              o[a]
            ) : /* @__PURE__ */ ns(i), c = i["~run"]({ value: u }, r);
            if (c.issues) {
              const l = {
                type: "object",
                origin: "value",
                input: o,
                key: a,
                value: u
              };
              for (const p of c.issues)
                p.path ? p.path.unshift(l) : p.path = [l], (s = n.issues) == null || s.push(p);
              if (n.issues || (n.issues = c.issues), r.abortEarly) {
                n.typed = !1;
                break;
              }
            }
            c.typed || (n.typed = !1), n.value[a] = c.value;
          } else if (i.type !== "exact_optional" && i.type !== "optional" && i.type !== "nullish" && (O(this, "key", n, r, {
            input: void 0,
            expected: `"${a}"`,
            path: [
              {
                type: "object",
                origin: "key",
                input: o,
                key: a,
                // @ts-expect-error
                value: o[a]
              }
            ]
          }), r.abortEarly))
            break;
        }
        if (!n.issues || !r.abortEarly)
          for (const a in o)
            /* @__PURE__ */ Xr(o, a) && !(a in this.entries) && (n.value[a] = o[a]);
      } else
        O(this, "type", n, r);
      return n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function et(e) {
  return {
    kind: "schema",
    type: "number",
    reference: et,
    expects: "number",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(t, n) {
      return typeof t.value == "number" && !isNaN(t.value) ? t.typed = !0 : O(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function v(e, t) {
  return {
    kind: "schema",
    type: "optional",
    reference: v,
    expects: `(${e.expects} | undefined)`,
    async: !1,
    wrapped: e,
    default: t,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(n, r) {
      return n.value === void 0 && (this.default !== void 0 && (n.value = /* @__PURE__ */ ns(this, n, r)), n.value === void 0) ? (n.typed = !0, n) : this.wrapped["~run"](n, r);
    }
  };
}
// @__NO_SIDE_EFFECTS__
function os(e, t, n) {
  return {
    kind: "schema",
    type: "record",
    reference: os,
    expects: "Object",
    async: !1,
    key: e,
    value: t,
    message: n,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(r, s) {
      var o, a;
      const i = r.value;
      if (i && typeof i == "object") {
        r.typed = !0, r.value = {};
        for (const u in i)
          if (/* @__PURE__ */ Xr(i, u)) {
            const c = i[u], l = this.key["~run"]({ value: u }, s);
            if (l.issues) {
              const m = {
                type: "object",
                origin: "key",
                input: i,
                key: u,
                value: c
              };
              for (const f of l.issues)
                f.path = [m], (o = r.issues) == null || o.push(f);
              if (r.issues || (r.issues = l.issues), s.abortEarly) {
                r.typed = !1;
                break;
              }
            }
            const p = this.value["~run"](
              { value: c },
              s
            );
            if (p.issues) {
              const m = {
                type: "object",
                origin: "value",
                input: i,
                key: u,
                value: c
              };
              for (const f of p.issues)
                f.path ? f.path.unshift(m) : f.path = [m], (a = r.issues) == null || a.push(f);
              if (r.issues || (r.issues = p.issues), s.abortEarly) {
                r.typed = !1;
                break;
              }
            }
            (!l.typed || !p.typed) && (r.typed = !1), l.typed && (r.value[l.value] = p.value);
          }
      } else
        O(this, "type", r, s);
      return r;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function _(e) {
  return {
    kind: "schema",
    type: "string",
    reference: _,
    expects: "string",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(t, n) {
      return typeof t.value == "string" ? t.typed = !0 : O(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function ir(e) {
  let t;
  if (e)
    for (const n of e)
      t ? t.push(...n.issues) : t = n.issues;
  return t;
}
// @__NO_SIDE_EFFECTS__
function nn(e, t) {
  return {
    kind: "schema",
    type: "union",
    reference: nn,
    expects: /* @__PURE__ */ Na(
      e.map((n) => n.expects),
      "|"
    ),
    async: !1,
    options: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(n, r) {
      let s, o, a;
      for (const i of this.options) {
        const u = i["~run"]({ value: n.value }, r);
        if (u.typed)
          if (u.issues)
            o ? o.push(u) : o = [u];
          else {
            s = u;
            break;
          }
        else
          a ? a.push(u) : a = [u];
      }
      if (s)
        return s;
      if (o) {
        if (o.length === 1)
          return o[0];
        O(this, "type", n, r, {
          issues: /* @__PURE__ */ ir(o)
        }), n.typed = !0;
      } else {
        if ((a == null ? void 0 : a.length) === 1)
          return a[0];
        O(this, "type", n, r, {
          issues: /* @__PURE__ */ ir(a)
        });
      }
      return n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function as() {
  return {
    kind: "schema",
    type: "unknown",
    reference: as,
    expects: "unknown",
    async: !1,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(e) {
      return e.typed = !0, e;
    }
  };
}
function is(e, t, n) {
  const r = e["~run"]({ value: t }, /* @__PURE__ */ Yr(n));
  if (r.issues)
    throw new Ha(r.issues);
  return r.value;
}
// @__NO_SIDE_EFFECTS__
function oe(...e) {
  return {
    ...e[0],
    pipe: e,
    get "~standard"() {
      return /* @__PURE__ */ F(this);
    },
    "~run"(t, n) {
      for (const r of e)
        if (r.kind !== "metadata") {
          if (t.issues && (r.kind === "schema" || r.kind === "transformation")) {
            t.typed = !1;
            break;
          }
          (!t.issues || !n.abortEarly && !n.abortPipeEarly) && (t = r["~run"](t, n));
        }
      return t;
    }
  };
}
function za(e) {
  return e.replace(/_[a-z]/g, (t) => t[1].toUpperCase());
}
function Ga(e) {
  return Object.entries(e).reduce((t, [n, r]) => (t[za(n)] = r, t), {});
}
function Dt(e) {
  const t = Ga(e);
  for (const n in t) {
    const r = t[n];
    r && typeof r == "object" && !(r instanceof Date) && (t[n] = Array.isArray(r) ? r.map(Dt) : Dt(r));
  }
  return t;
}
function rn(e) {
  return /* @__PURE__ */ se((t) => e ? Dt(t) : t);
}
function us(e) {
  return (t) => /* @__PURE__ */ oe(
    e,
    rn(t)
  );
}
function Qa(e) {
  return (t, n) => is(
    /* @__PURE__ */ oe(e, rn(n)),
    t
  );
}
function sn() {
  return /* @__PURE__ */ se(JSON.parse);
}
function on(e) {
  const t = us(e);
  return (n) => /* @__PURE__ */ oe(
    /* @__PURE__ */ _(),
    sn(),
    t(n)
  );
}
function Ja(e) {
  return /* @__PURE__ */ se((t) => {
    const n = {};
    return new URLSearchParams(t).forEach((r, s) => {
      const o = n[s];
      Array.isArray(o) ? o.push(r) : o === void 0 ? n[s] = r : n[s] = [o, r];
    }), is(e, n);
  });
}
function cs(e) {
  return (t) => /* @__PURE__ */ oe(
    /* @__PURE__ */ nn([/* @__PURE__ */ _(), /* @__PURE__ */ ss(URLSearchParams)]),
    Ja(e),
    rn(t)
  );
}
const ur = /* @__PURE__ */ v(/* @__PURE__ */ tn(() => ei())), Ka = /* @__PURE__ */ Se({
  id: /* @__PURE__ */ et(),
  photo_url: /* @__PURE__ */ v(/* @__PURE__ */ _()),
  type: /* @__PURE__ */ _(),
  title: /* @__PURE__ */ _(),
  username: /* @__PURE__ */ v(/* @__PURE__ */ _())
}), Za = /* @__PURE__ */ Se({
  added_to_attachment_menu: /* @__PURE__ */ v(/* @__PURE__ */ je()),
  allows_write_to_pm: /* @__PURE__ */ v(/* @__PURE__ */ je()),
  first_name: /* @__PURE__ */ _(),
  id: /* @__PURE__ */ et(),
  is_bot: /* @__PURE__ */ v(/* @__PURE__ */ je()),
  is_premium: /* @__PURE__ */ v(/* @__PURE__ */ je()),
  last_name: /* @__PURE__ */ v(/* @__PURE__ */ _()),
  language_code: /* @__PURE__ */ v(/* @__PURE__ */ _()),
  photo_url: /* @__PURE__ */ v(/* @__PURE__ */ _()),
  username: /* @__PURE__ */ v(/* @__PURE__ */ _())
}), Ya = /* @__PURE__ */ Se({
  auth_date: /* @__PURE__ */ oe(
    /* @__PURE__ */ _(),
    /* @__PURE__ */ se((e) => new Date(Number(e) * 1e3)),
    /* @__PURE__ */ rs()
  ),
  can_send_after: /* @__PURE__ */ v(/* @__PURE__ */ oe(/* @__PURE__ */ _(), /* @__PURE__ */ se(Number), /* @__PURE__ */ ts())),
  chat: /* @__PURE__ */ v(/* @__PURE__ */ tn(() => Xa())),
  chat_type: /* @__PURE__ */ v(/* @__PURE__ */ _()),
  chat_instance: /* @__PURE__ */ v(/* @__PURE__ */ _()),
  hash: /* @__PURE__ */ _(),
  query_id: /* @__PURE__ */ v(/* @__PURE__ */ _()),
  receiver: ur,
  start_param: /* @__PURE__ */ v(/* @__PURE__ */ _()),
  signature: /* @__PURE__ */ _(),
  user: ur
}), Xa = on(Ka), ei = on(Za), ti = cs(Ya);
function ni(e) {
  return /^#[\da-f]{6}$/i.test(e);
}
const ls = us(
  /* @__PURE__ */ os(
    /* @__PURE__ */ _(),
    /* @__PURE__ */ oe(
      /* @__PURE__ */ nn([/* @__PURE__ */ _(), /* @__PURE__ */ et()]),
      /* @__PURE__ */ se((e) => typeof e == "number" ? `#${(e & 16777215).toString(16).padStart(6, "0")}` : e),
      /* @__PURE__ */ es(ni)
    )
  )
), kt = /* @__PURE__ */ v(
  /* @__PURE__ */ oe(/* @__PURE__ */ _(), /* @__PURE__ */ se((e) => e === "1"))
), cr = on(ls()), ri = /* @__PURE__ */ Se({
  tgWebAppBotInline: kt,
  tgWebAppData: /* @__PURE__ */ v(ti()),
  tgWebAppDefaultColors: /* @__PURE__ */ v(cr()),
  tgWebAppFullscreen: kt,
  tgWebAppPlatform: /* @__PURE__ */ _(),
  tgWebAppShowSettings: kt,
  tgWebAppStartParam: /* @__PURE__ */ v(/* @__PURE__ */ _()),
  tgWebAppThemeParams: cr(),
  tgWebAppVersion: /* @__PURE__ */ _()
}), ps = cs(ri), si = Qa(ps()), ds = /* @__PURE__ */ Se({
  eventType: /* @__PURE__ */ _(),
  eventData: /* @__PURE__ */ v(/* @__PURE__ */ as())
});
function fs(e, t) {
  return t || (t = (n, r) => JSON.stringify(r)), new URLSearchParams(
    Object.entries(e).reduce((n, [r, s]) => (Array.isArray(s) ? n.push(...s.map((o) => [r, String(o)])) : s != null && n.push([
      r,
      s instanceof Date ? (s.getTime() / 1e3 | 0).toString() : typeof s == "string" || typeof s == "number" ? String(s) : typeof s == "boolean" ? s ? "1" : "0" : t(r, s)
    ]), n), [])
  ).toString();
}
function oi(e) {
  return fs(e);
}
function ai(e) {
  return fs(e, (t, n) => t === "tgWebAppData" ? oi(n) : JSON.stringify(n));
}
function ms(e) {
  try {
    return /* @__PURE__ */ Ua(ps(), e);
  } catch {
    return !1;
  }
}
function ii(e) {
  return { all: e = e || /* @__PURE__ */ new Map(), on: function(t, n) {
    var r = e.get(t);
    r ? r.push(n) : e.set(t, [n]);
  }, off: function(t, n) {
    var r = e.get(t);
    r && (n ? r.splice(r.indexOf(n) >>> 0, 1) : e.set(t, []));
  }, emit: function(t, n) {
    var r = e.get(t);
    r && r.slice().map(function(s) {
      s(n);
    }), (r = e.get("*")) && r.slice().map(function(s) {
      s(t, n);
    });
  } };
}
function ui(e, t) {
  const n = ii(), r = /* @__PURE__ */ new Map(), s = (o, a, i) => {
    i || (i = !1);
    const u = r.get(o) || /* @__PURE__ */ new Map();
    r.set(o, u);
    const c = u.get(a) || [];
    u.set(a, c);
    const l = c.findIndex((p) => p[1] === i);
    l >= 0 && (n.off(o, c[l][0]), c.splice(l, 1), !c.length && u.delete(a), u.size || (r.delete(o), !r.size && t()));
  };
  return [
    function(o, a, i) {
      !r.size && e();
      function u() {
        s(o, a, i);
      }
      function c(...m) {
        i && u(), o === "*" ? a(m) : a(...m);
      }
      n.on(o, c);
      const l = r.get(o) || /* @__PURE__ */ new Map();
      r.set(o, l);
      const p = l.get(a) || [];
      return l.set(a, p), p.push([c, i || !1]), u;
    },
    s,
    // eslint-disable-next-line @typescript-eslint/unbound-method
    n.emit,
    function() {
      const o = n.all.size;
      n.all.clear(), r.clear(), o && t();
    }
  ];
}
function lr(e, t) {
  window.dispatchEvent(new MessageEvent("message", {
    data: JSON.stringify({ eventType: e, eventData: t }),
    // We specify window.parent to imitate the case, the parent iframe sent us this event.
    source: window.parent
  }));
}
let He = !1;
const [an, ci] = /* @__PURE__ */ Va("Bridge", {
  bgColor: "#9147ff",
  textColor: "white",
  shouldLog() {
    return He;
  }
}), pr = (e) => {
  an(!1, "Event received:", e);
};
function li(e) {
  e !== He && (He = e, He ? C("*", pr) : M("*", pr));
}
const pi = {
  clipboard_text_received: /* @__PURE__ */ N({
    req_id: /* @__PURE__ */ ne(),
    data: /* @__PURE__ */ Be(/* @__PURE__ */ ne())
  }),
  custom_method_invoked: /* @__PURE__ */ N({
    req_id: /* @__PURE__ */ ne(),
    result: /* @__PURE__ */ $t(/* @__PURE__ */ Nr()),
    error: /* @__PURE__ */ $t(/* @__PURE__ */ ne())
  }),
  popup_closed: /* @__PURE__ */ Be(
    /* @__PURE__ */ N({ button_id: /* @__PURE__ */ Be(/* @__PURE__ */ ne(), () => {
    }) }),
    {}
  ),
  viewport_changed: /* @__PURE__ */ N({
    height: /* @__PURE__ */ Tt(),
    width: /* @__PURE__ */ Be(/* @__PURE__ */ Tt(), () => window.innerWidth),
    is_state_stable: /* @__PURE__ */ Mt(),
    is_expanded: /* @__PURE__ */ Mt()
  }),
  theme_changed: /* @__PURE__ */ N({
    theme_params: ls()
  })
};
function dr(e) {
  if (e.source !== window.parent)
    return;
  let t;
  try {
    t = It(/* @__PURE__ */ Hr(/* @__PURE__ */ ne(), sn(), ds), e.data);
  } catch {
    return;
  }
  const { eventType: n, eventData: r } = t, s = pi[n];
  try {
    const o = s ? It(s, r) : r;
    di(n, o);
  } catch (o) {
    ci(
      !0,
      [
        `An error occurred processing the "${n}" event from the Telegram application.`,
        "Please, file an issue here:",
        "https://github.com/Telegram-Mini-Apps/telegram-apps/issues/new/choose"
      ].join(`
`),
      t,
      o
    );
  }
}
const [
  C,
  M,
  di,
  cf
] = ui(
  () => {
    const e = window, t = { receiveEvent: lr };
    e.TelegramGameProxy_receiveEvent = lr, e.TelegramGameProxy = t, e.Telegram = { WebView: t }, window.addEventListener("message", dr);
  },
  () => {
    ["TelegramGameProxy_receiveEvent", "TelegramGameProxy", "Telegram"].forEach((e) => {
      delete window[e];
    }), window.removeEventListener("message", dr);
  }
);
function fi(e) {
  return (t) => t instanceof e;
}
function ke(e, t) {
  t || (t = []);
  class n extends Error {
    constructor(...s) {
      const o = typeof t == "function" ? t(...s) : typeof t == "string" ? [t] : t || [];
      super(...o), this.name = e;
    }
  }
  return Object.defineProperty(n, "name", { value: e }), [n, fi(n)];
}
const [
  mi,
  lf
] = ke(
  "MethodUnsupportedError",
  (e, t) => [
    `Method "${e}" is unsupported in Mini Apps version ${t}`
  ]
), [
  hi,
  pf
] = ke(
  "MethodParameterUnsupportedError",
  (e, t, n) => [
    `Parameter "${t}" of "${e}" method is unsupported in Mini Apps version ${n}`
  ]
), bi = [
  "Unable to retrieve launch parameters from any known source. Perhaps, you have opened your app outside Telegram?",
  "📖 Refer to docs for more information:",
  "https://docs.telegram-mini-apps.com/packages/telegram-apps-bridge/environment"
].join(`
`), [
  yi,
  df
] = ke("LaunchParamsRetrieveError", bi), [
  _i,
  ff
] = ke("InvalidLaunchParamsError", (e) => [
  `Invalid value for launch params: ${e}`
]), [gi, mf] = ke("UnknownEnvError"), [
  vi,
  hf
] = ke(
  "InvokeCustomMethodError",
  (e) => [`Server returned error: ${e}`]
);
function wi(e, t) {
  t();
}
// @__NO_SIDE_EFFECTS__
function Si(e, t) {
  t || (t = {});
  const n = t.equals || Object.is;
  let r = [], s = e;
  const o = (c) => {
    if (!n(s, c)) {
      const l = s;
      s = c, wi(u, () => {
        [...r].forEach(([p, m]) => {
          p(c, l), m && i(p, !0);
        });
      });
    }
  };
  function a(c) {
    const l = typeof c != "object" ? { once: c } : c;
    return {
      once: l.once || !1,
      signal: l.signal || !1
    };
  }
  const i = (c, l) => {
    const p = a(l), m = r.findIndex(([f, x]) => f === c && x.once === p.once && x.signal === p.signal);
    m >= 0 && r.splice(m, 1);
  }, u = Object.assign(
    function() {
      return ki(u), s;
    },
    {
      destroy() {
        r = [];
      },
      set: o,
      reset() {
        o(e);
      },
      sub(c, l) {
        return r.push([c, a(l)]), () => i(c, l);
      },
      unsub: i,
      unsubAll() {
        r = r.filter((c) => c[1].signal);
      }
    }
  );
  return u;
}
const Et = [];
function ki(e) {
  Et.length && Et[Et.length - 1].add(e);
}
const Ei = /* @__PURE__ */ Si("https://web.telegram.org");
function un(e, t) {
  an(!1, "Posting event:", t ? { eventType: e, eventData: t } : { eventType: e });
  const n = window, r = JSON.stringify({ eventType: e, eventData: t });
  if (zr())
    return n.parent.postMessage(r, Ei());
  if (Ur(n)) {
    n.TelegramWebviewProxy.postEvent(e, JSON.stringify(t));
    return;
  }
  if (/* @__PURE__ */ Xe(/* @__PURE__ */ N({ external: /* @__PURE__ */ N({ notify: /* @__PURE__ */ en() }) }), n)) {
    n.external.notify(r);
    return;
  }
  throw new gi();
}
function cn(e, t, n) {
  n || (n = {});
  const { capture: r } = n, [s, o] = Ra();
  return new Jr((a) => {
    (Array.isArray(t) ? t : [t]).forEach((i) => {
      s(
        C(i, (u) => {
          (!r || (Array.isArray(t) ? r({
            event: i,
            payload: u
          }) : r(u))) && a(u);
        })
      );
    }), (n.postEvent || un)(e, n.params);
  }, n).finally(o);
}
const fr = "launchParams";
function mr(e) {
  return e.replace(/^[^?#]*[?#]/, "").replace(/[?#]/g, "&");
}
function hs() {
  for (const e of [
    // Try to retrieve launch parameters from the current location. This method can return
    // nothing in case, location was changed, and then the page was reloaded.
    () => mr(window.location.href),
    // Then, try using the lower level API - window.performance.
    () => {
      const t = performance.getEntriesByType("navigation")[0];
      return t && mr(t.name);
    },
    () => Da(fr)
  ]) {
    const t = e();
    if (t && ms(t))
      return Zr(fr, t), t;
  }
  throw new yi();
}
function ln(e) {
  const t = si(hs());
  return e ? Lt(t) : t;
}
function hr(e, t) {
  if (!e)
    try {
      return ln(), !0;
    } catch {
      return !1;
    }
  return Jr.fn(async (n) => {
    if (Ur(window))
      return !0;
    try {
      return await cn("web_app_request_theme", "theme_changed", n), !0;
    } catch {
      return !1;
    }
  }, t || { timeout: 100 });
}
function bf({ launchParams: e, onEvent: t } = {}) {
  if (e) {
    const s = typeof e == "string" || e instanceof URLSearchParams ? e.toString() : (
      // Here we have to trick serializeLaunchParamsQuery into thinking, it serializes a valid
      // value. We are doing it because we are working with tgWebAppData presented as a
      // string, not an object as serializeLaunchParamsQuery requires.
      ai({ ...e, tgWebAppData: void 0 }) + (e.tgWebAppData ? `&tgWebAppData=${encodeURIComponent(e.tgWebAppData.toString())}` : "")
    );
    if (!ms(s))
      throw new _i(s);
    Zr("launchParams", s);
  }
  if (zr()) {
    const s = /* @__PURE__ */ Hr(
      /* @__PURE__ */ ne(),
      sn(),
      ds
    ), o = window.parent.postMessage.bind(window.parent);
    window.parent.postMessage = (...a) => {
      const [i] = a, u = () => {
        o(...a);
      };
      if (/* @__PURE__ */ Xe(s, i) && t) {
        const c = It(s, i);
        t([c.eventType, c.eventData], u);
      } else
        u();
    };
    return;
  }
  const n = window.TelegramWebviewProxy || {}, { postEvent: r } = n;
  window.TelegramWebviewProxy = {
    ...n,
    postEvent(s, o) {
      const a = () => {
        r && r(s, o);
      };
      t ? t([s, o ? JSON.parse(o) : void 0], a) : a();
    }
  }, an(!1, "Environment was mocked by the mockTelegramEnv function");
}
function Ai() {
  return new URLSearchParams(hs()).get("tgWebAppData") || void 0;
}
function bs(e) {
  return ({ req_id: t }) => t === e;
}
function br(e) {
  return e.split(".").map(Number);
}
function Ci(e, t) {
  const n = br(e), r = br(t), s = Math.max(n.length, r.length);
  for (let o = 0; o < s; o += 1) {
    const a = n[o] || 0, i = r[o] || 0;
    if (a !== i)
      return a > i ? 1 : -1;
  }
  return 0;
}
function B(e, t) {
  return Ci(e, t) <= 0;
}
function pe(e, t, n) {
  if (typeof n == "string") {
    if (e === "web_app_open_link") {
      if (t === "try_instant_view")
        return B("6.4", n);
      if (t === "try_browser")
        return B("7.6", n);
    }
    if (e === "web_app_set_header_color" && t === "color")
      return B("6.9", n);
    if (e === "web_app_close" && t === "return_back")
      return B("7.6", n);
    if (e === "web_app_setup_main_button" && t === "has_shine_effect")
      return B("7.10", n);
  }
  switch (e) {
    case "web_app_open_tg_link":
    case "web_app_open_invoice":
    case "web_app_setup_back_button":
    case "web_app_set_background_color":
    case "web_app_set_header_color":
    case "web_app_trigger_haptic_feedback":
      return B("6.1", t);
    case "web_app_open_popup":
      return B("6.2", t);
    case "web_app_close_scan_qr_popup":
    case "web_app_open_scan_qr_popup":
    case "web_app_read_text_from_clipboard":
      return B("6.4", t);
    case "web_app_switch_inline_query":
      return B("6.7", t);
    case "web_app_invoke_custom_method":
    case "web_app_request_write_access":
    case "web_app_request_phone":
      return B("6.9", t);
    case "web_app_setup_settings_button":
      return B("6.10", t);
    case "web_app_biometry_get_info":
    case "web_app_biometry_open_settings":
    case "web_app_biometry_request_access":
    case "web_app_biometry_request_auth":
    case "web_app_biometry_update_token":
      return B("7.2", t);
    case "web_app_setup_swipe_behavior":
      return B("7.7", t);
    case "web_app_share_to_story":
      return B("7.8", t);
    case "web_app_setup_secondary_button":
    case "web_app_set_bottom_bar_color":
      return B("7.10", t);
    case "web_app_request_safe_area":
    case "web_app_request_content_safe_area":
    case "web_app_request_fullscreen":
    case "web_app_exit_fullscreen":
    case "web_app_set_emoji_status":
    case "web_app_add_to_home_screen":
    case "web_app_check_home_screen":
    case "web_app_request_emoji_status_access":
    case "web_app_check_location":
    case "web_app_open_location_settings":
    case "web_app_request_file_download":
    case "web_app_request_location":
    case "web_app_send_prepared_message":
    case "web_app_start_accelerometer":
    case "web_app_start_device_orientation":
    case "web_app_start_gyroscope":
    case "web_app_stop_accelerometer":
    case "web_app_stop_device_orientation":
    case "web_app_stop_gyroscope":
    case "web_app_toggle_orientation_lock":
      return B("8.0", t);
    default:
      return [
        "iframe_ready",
        "iframe_will_reload",
        "web_app_close",
        "web_app_data_send",
        "web_app_expand",
        "web_app_open_link",
        "web_app_ready",
        "web_app_request_theme",
        "web_app_request_viewport",
        "web_app_setup_main_button",
        "web_app_setup_closing_behavior"
      ].includes(e);
  }
}
function xi(e, t) {
  t || (t = "strict");
  const n = typeof t == "function" ? t : (r) => {
    const { method: s, version: o } = r, a = "param" in r ? new hi(s, r.param, o) : new mi(s, o);
    if (t === "strict")
      throw a;
    return console.warn(a.message);
  };
  return (r, s) => pe(r, e) ? r === "web_app_set_header_color" && /* @__PURE__ */ Xe(/* @__PURE__ */ N({ color: /* @__PURE__ */ Wr() }), s) && !pe(r, "color", e) ? n({ version: e, method: r, param: "color" }) : un(r, s) : n({ version: e, method: r });
}
function Pi(e, t, n, r) {
  return cn("web_app_invoke_custom_method", "custom_method_invoked", {
    ...r || {},
    params: { method: e, params: t, req_id: n },
    capture: bs(n)
  }).then(({ result: s, error: o }) => {
    if (o)
      throw new vi(o);
    return s;
  });
}
function ys(e) {
  return e.replace(/[A-Z]/g, (t) => `-${t.toLowerCase()}`);
}
function Bi(e) {
  return e.replace(/_([a-z])/g, (t, n) => `-${n.toLowerCase()}`);
}
function _s(e) {
  return `tapps/${e}`;
}
function H(e, t) {
  sessionStorage.setItem(_s(e), JSON.stringify(t));
}
function U(e) {
  const t = sessionStorage.getItem(_s(e));
  try {
    return t ? JSON.parse(t) : void 0;
  } catch {
  }
}
function pn(...e) {
  const t = e.flat(1);
  return [
    t.push.bind(t),
    () => {
      t.forEach((n) => {
        n();
      });
    }
  ];
}
// @__NO_SIDE_EFFECTS__
function ji(e, t) {
  t || (t = {});
  const {
    textColor: n,
    bgColor: r,
    shouldLog: s
  } = t, o = s === void 0 ? !0 : s, a = typeof o == "boolean" ? () => o : o;
  function i(u, c, ...l) {
    if (c || a()) {
      const p = "font-weight:bold;padding:0 5px;border-radius:5px";
      console[u](
        `%c${Intl.DateTimeFormat("en-GB", {
          hour: "2-digit",
          minute: "2-digit",
          second: "2-digit",
          fractionalSecondDigits: 3,
          timeZone: "UTC"
        }).format(/* @__PURE__ */ new Date())}%c / %c${e}`,
        `${p};background-color: lightblue;color:black`,
        "",
        `${p};${n ? `color:${n};` : ""}${r ? `background-color:${r}` : ""}`,
        ...l
      );
    }
  }
  return [i.bind(void 0, "log"), i.bind(void 0, "error")];
}
function qi() {
  return performance.getEntriesByType("navigation")[0];
}
function z() {
  const e = qi();
  return !!e && e.type === "reload";
}
let ye;
function Mi(e, t) {
  ye && ye.set(e, t) || t();
}
function Rt(e) {
  if (ye)
    return e();
  ye = /* @__PURE__ */ new Map();
  try {
    e();
  } finally {
    ye.forEach((t) => t()), ye = void 0;
  }
}
// @__NO_SIDE_EFFECTS__
function gs(e, t) {
  t || (t = {});
  const n = t.equals || Object.is;
  let r = [], s = e;
  const o = (c) => {
    if (!n(s, c)) {
      const l = s;
      s = c, Mi(u, () => {
        [...r].forEach(([p, m]) => {
          p(c, l), m && i(p, !0);
        });
      });
    }
  };
  function a(c) {
    const l = typeof c != "object" ? { once: c } : c;
    return {
      once: l.once || !1,
      signal: l.signal || !1
    };
  }
  const i = (c, l) => {
    const p = a(l), m = r.findIndex(([f, x]) => f === c && x.once === p.once && x.signal === p.signal);
    m >= 0 && r.splice(m, 1);
  }, u = Object.assign(
    function() {
      return Ti(u), s;
    },
    {
      destroy() {
        r = [];
      },
      set: o,
      reset() {
        o(e);
      },
      sub(c, l) {
        return r.push([c, a(l)]), () => i(c, l);
      },
      unsub: i,
      unsubAll() {
        r = r.filter((c) => c[1].signal);
      }
    }
  );
  return u;
}
const Me = [];
function Ti(e) {
  Me.length && Me[Me.length - 1].add(e);
}
// @__NO_SIDE_EFFECTS__
function $i(e, t) {
  let n = /* @__PURE__ */ new Set(), r;
  function s() {
    return r || (r = /* @__PURE__ */ gs(a(), t));
  }
  function o() {
    s().set(a());
  }
  function a() {
    n.forEach((c) => {
      c.unsub(o, { signal: !0 });
    });
    const i = /* @__PURE__ */ new Set();
    let u;
    Me.push(i);
    try {
      u = e();
    } finally {
      Me.pop();
    }
    return i.forEach((c) => {
      c.sub(o, { signal: !0 });
    }), n = i, u;
  }
  return Object.assign(function() {
    return s()();
  }, {
    destroy() {
      s().destroy();
    },
    sub(...i) {
      return s().sub(...i);
    },
    unsub(...i) {
      s().unsub(...i);
    },
    unsubAll(...i) {
      s().unsubAll(...i);
    }
  });
}
// @__NO_SIDE_EFFECTS__
function Ee(e, t) {
  return /* @__PURE__ */ gs(e, t);
}
// @__NO_SIDE_EFFECTS__
function h(e, t) {
  return /* @__PURE__ */ $i(e, t);
}
// @__NO_SIDE_EFFECTS__
function y(e, t) {
  const n = /* @__PURE__ */ Ee(e, t);
  return [n, /* @__PURE__ */ h(n)];
}
let vs = !1;
const [Vt, yf] = /* @__PURE__ */ ji("SDK", {
  bgColor: "forestgreen",
  textColor: "white",
  shouldLog() {
    return vs;
  }
});
function _f(e) {
  vs = e, li(e);
}
const At = /* @__PURE__ */ Ee(0), ws = /* @__PURE__ */ Ee(un), [yr, dn] = /* @__PURE__ */ y({
  tgWebAppPlatform: "unknown",
  tgWebAppVersion: "0.0"
}), re = /* @__PURE__ */ h(() => dn().tgWebAppVersion);
function Ii(e) {
  e || (e = {});
  const { postEvent: t } = e, n = e.launchParams || ln();
  yr.set(n), ws.set(
    typeof t == "function" ? t : xi(n.tgWebAppVersion)
  ), Vt(!1, "The package was configured. Launch params:", yr());
}
function Ss() {
  return At.set(At() + 1), At().toString();
}
function Ae(e, t, n) {
  return Pi(e, t, Ss(), {
    ...n || {},
    postEvent: b
  });
}
const E = (e, t, n) => (n || (n = {}), n.postEvent || (n.postEvent = b), cn(e, t, n)), b = (e, t) => ws()(e, t);
function G(e) {
  return /* @__PURE__ */ h(() => pe(e, re()));
}
function Oi(e) {
  return (t) => t instanceof e;
}
function Z(e, t) {
  t || (t = []);
  class n extends Error {
    constructor(...s) {
      const o = typeof t == "function" ? t(...s) : typeof t == "string" ? [t] : t || [];
      super(...o), this.name = e;
    }
  }
  return Object.defineProperty(n, "name", { value: e }), [n, Oi(n)];
}
function de(e) {
  return [e];
}
const [
  fn,
  gf
] = Z("CSSVarsBoundError", "CSS variables are already bound"), [
  ks,
  vf
] = Z("NotAvailableError", de), [
  wf,
  Sf
] = Z("InvalidEnvError", de), [
  Pe,
  kf
] = Z("FunctionNotAvailableError", de), [
  J,
  Ef
] = Z(
  "InvalidArgumentsError",
  (e, t) => [e, { cause: t }]
), [
  Li,
  Af
] = Z("ConcurrentCallError", de), [
  Di,
  Cf
] = Z(
  "SetEmojiStatusError",
  (e) => [`Failed to set emoji status: ${e}`]
), [
  Es,
  xf
] = Z("AccessDeniedError", de), [
  Ri,
  Pf
] = Z("FullscreenFailedError", de), [
  Vi,
  Bf
] = Z("ShareMessageError", de);
function _r() {
  return typeof window > "u";
}
// @__NO_SIDE_EFFECTS__
function A(e, t, n) {
  n || (n = {});
  const {
    isSupported: r,
    isMounted: s,
    isMounting: o,
    component: a,
    supports: i
  } = n || {}, u = `${a ? `${a}.` : ""}${e}()`, c = r ? Array.isArray(r) || typeof r == "object" && "any" in r ? r : [r] : void 0;
  function l(S) {
    if (i) {
      const k = i[S];
      return pe(k[0], k[1], re());
    }
    return !0;
  }
  function p() {
    if (!c)
      return;
    function S(V) {
      return typeof V == "function" ? V() : pe(V, re()) ? void 0 : `it is unsupported in Mini Apps version ${re()}`;
    }
    const k = Array.isArray(c) ? c : c.any, $ = k.map(S).filter(Boolean);
    return Array.isArray(c) ? $[0] : $.length === k.length ? $[$.length - 1] : void 0;
  }
  function m(...S) {
    for (const k in i)
      if (i[k][2](...S) && !l(k))
        return `option ${k} is not supported in Mini Apps version ${re()}`;
  }
  let f;
  if (i) {
    f = {};
    for (const S in i)
      f[S] = /* @__PURE__ */ h(() => l(S));
  }
  const x = /* @__PURE__ */ h(() => !p()), be = /* @__PURE__ */ h(() => re() !== "0.0"), P = /* @__PURE__ */ h(() => !s || s()), te = /* @__PURE__ */ h(
    () => hr() && !_r() && be() && x() && P()
  );
  return Object.assign(
    (...S) => {
      const k = `Unable to call the ${u} ${a ? "method" : "function"}:`;
      if (_r() || !hr())
        throw new Pe(`${k} it can't be called outside Mini Apps`);
      if (!be())
        throw new Pe(`${k} the SDK was not initialized. Use the SDK init() function`);
      const $ = p();
      if ($)
        throw new Pe(`${k} ${$}`);
      const V = m(...S);
      if (V)
        throw new Pe(`${k} ${V}`);
      if (!P()) {
        const d = o && o() ? "mounting. Wait for the mount completion" : `unmounted. Use the ${a}.mount() method`;
        throw new Pe(`${k} the component is ${d}`);
      }
      return t(...S);
    },
    t,
    {
      isAvailable: te,
      ifAvailable(...S) {
        return te() ? [!0, t(...S)] : [!1];
      }
    },
    c ? { isSupported: x } : {},
    f ? { supports: f } : {}
  );
}
function tt(e, t) {
  return t || (t = {}), (n, r, s, o) => /* @__PURE__ */ A(n, r, {
    ...t,
    isSupported: s || t.isSupported,
    supports: o,
    component: e
  });
}
function ue(e, t, n) {
  return tt(e, { isSupported: n, isMounted: t });
}
function R(e, t) {
  return tt(e, { isSupported: t });
}
const nt = "web_app_setup_back_button", As = "back_button_pressed", rt = "backButton", [gr, Fi] = /* @__PURE__ */ y(!1), [ze, Wi] = /* @__PURE__ */ y(!1), Ni = G(nt), Cs = ue(rt, ze, nt), mn = R(rt, nt), Hi = Cs("hide", () => {
  hn(!1);
}), Ui = mn("mount", () => {
  ze() || (hn(z() && U(rt) || !1), ze.set(!0));
});
function hn(e) {
  e !== gr() && (b(nt, { is_visible: e }), H(rt, e), gr.set(e));
}
const zi = mn(
  "onClick",
  (e) => C(As, e)
), Gi = mn(
  "offClick",
  (e) => {
    M(As, e);
  }
), Qi = Cs("show", () => {
  hn(!0);
});
function Ji() {
  ze.set(!1);
}
const jf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  hide: Hi,
  isMounted: Wi,
  isSupported: Ni,
  isVisible: Fi,
  mount: Ui,
  offClick: Gi,
  onClick: zi,
  show: Qi,
  unmount: Ji
}, Symbol.toStringTag, { value: "Module" }));
var Ki = Object.defineProperty, Zi = (e, t, n) => t in e ? Ki(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n, Ft = (e, t, n) => Zi(e, typeof t != "symbol" ? t + "" : t, n);
function Yi(e) {
  return (t) => t instanceof e;
}
function xs(e, t) {
  t || (t = []);
  class n extends Error {
    constructor(...s) {
      const o = typeof t == "function" ? t(...s) : typeof t == "string" ? [t] : t || [];
      super(...o), this.name = e;
    }
  }
  return Object.defineProperty(n, "name", { value: e }), [n, Yi(n)];
}
const [Xi, eu] = xs("CancelledError", "Promise was canceled"), [tu, qf] = xs(
  "TimeoutError",
  (e, t) => [`Timeout reached: ${e}ms`, { cause: t }]
), Ps = Symbol("Resolved");
function vr(e) {
  return Array.isArray(e) && e[0] === Ps;
}
function nu(e) {
  return [Ps, e];
}
function wr(e, t) {
  return e.reject = t.reject, e.abort = t.abort, e;
}
let j = class Wt extends Promise {
  constructor(t, n) {
    let r, s;
    super((o, a) => {
      let i, u;
      typeof t == "function" ? (i = t, u = n) : u = t;
      const c = [], l = (d) => (...I) => {
        const ee = d(...I);
        return c.forEach((St) => St()), ee;
      }, p = new AbortController(), { signal: m } = p;
      s = (d) => {
        !m.aborted && p.abort(d);
      };
      const f = () => m.reason, x = (d) => {
        const I = () => {
          d(f());
        };
        m.addEventListener("abort", I, !0);
        const ee = () => {
          m.removeEventListener("abort", I, !0);
        };
        return c.push(ee), ee;
      }, be = l((d) => {
        o(d), s(nu(d));
      });
      r = l((d) => {
        a(d), s(d);
      }), u || (u = {});
      const { abortSignal: P, rejectOnAbort: te = !0 } = u;
      if (P)
        if (P.aborted) {
          const { reason: d } = P;
          if (te)
            return r(d);
          s(d);
        } else {
          const d = () => {
            s(P.reason);
          };
          P.addEventListener("abort", d), c.push(() => {
            P.removeEventListener("abort", d);
          });
        }
      te && x(a);
      const { timeout: S } = u;
      if (S) {
        const d = setTimeout(() => {
          s(new tu(S));
        }, S);
        c.push(() => {
          clearTimeout(d);
        });
      }
      const k = () => m.aborted, $ = () => vr(f()), V = () => {
        const d = f();
        return vr(d) ? d[1] : void 0;
      };
      try {
        const d = i && i(be, r, {
          abortReason: f,
          abortSignal: m,
          isAborted: k,
          isResolved: $,
          onAborted: x,
          onResolved: (I) => x(() => {
            $() && I(V());
          }),
          resolved: V,
          throwIfAborted() {
            if (k())
              throw f();
          }
        });
        d instanceof Promise && d.catch(r);
      } catch (d) {
        r(d);
      }
    }), Ft(this, "abort"), Ft(this, "reject"), this.abort = s, this.reject = r;
  }
  /**
   * Creates a new AbortablePromise instance using an executor, resolving the promise when a result
   * was returned.
   * @param fn - function returning promise result.
   * @param options - additional options.
   */
  static fn(t, n) {
    return new Wt(async (r, s, o) => {
      try {
        r(await t(o));
      } catch (a) {
        s(a);
      }
    }, n);
  }
  static resolve(t) {
    return this.fn(() => t);
  }
  /**
   * @see Promise.reject
   */
  static reject(t) {
    return new Wt((n, r) => {
      r(t);
    });
  }
  /**
   * Aborts the promise with the cancel error.
   */
  cancel() {
    this.abort(new Xi());
  }
  /**
   * @see Promise.catch
   */
  catch(t) {
    return this.then(void 0, t);
  }
  /**
   * @see Promise.finally
   */
  finally(t) {
    return wr(super.finally(t), this);
  }
  /**
   * @see Promise.then
   */
  then(t, n) {
    return wr(super.then(t, n), this);
  }
};
function Sr(e, t) {
  return e.resolve = t.resolve, e;
}
let ru = class Nt extends j {
  constructor(t, n) {
    let r, s;
    typeof t == "function" ? (r = t, s = n) : s = t;
    let o;
    super((a, i, u) => {
      o = a, r && r(a, i, u);
    }, s), Ft(this, "resolve"), this.resolve = o;
  }
  /**
   * Creates a new ManualPromise instance using an executor, resolving the promise when a result
   * was returned.
   * @param fn - function returning promise result.
   * @param options - additional options.
   */
  static fn(t, n) {
    return new Nt((r, s, o) => {
      try {
        Promise.resolve(t(o)).then(r, s);
      } catch (a) {
        s(a);
      }
    }, n);
  }
  static resolve(t) {
    return this.fn(() => t);
  }
  /**
   * @see Promise.reject
   */
  static reject(t) {
    return new Nt((n, r) => {
      r(t);
    });
  }
  /**
   * @see Promise.catch
   */
  catch(t) {
    return this.then(void 0, t);
  }
  /**
   * @see Promise.finally
   */
  finally(t) {
    return Sr(super.finally(t), this);
  }
  /**
   * @see Promise.then
   */
  then(t, n) {
    return Sr(super.then(t, n), this);
  }
};
function T(e, t, n) {
  n || (n = {});
  const {
    promise: r,
    error: s
  } = n, [o, a] = r ? [r, /* @__PURE__ */ h(r)] : /* @__PURE__ */ y(), [i, u] = s ? [s, /* @__PURE__ */ h(s)] : /* @__PURE__ */ y();
  return [
    Object.assign((...c) => {
      if (o()) {
        const p = new Li(t);
        return i.set(p), j.reject(p);
      }
      Rt(() => {
        o.set(e(...c)), i.set(void 0);
      });
      let l;
      return o().catch((p) => {
        throw l = p, p;
      }).finally(() => {
        Rt(() => {
          o.set(void 0), i.set(l);
        });
      });
    }, e),
    [o, a, /* @__PURE__ */ h(() => !!o())],
    [i, u]
  ];
}
// @__NO_SIDE_EFFECTS__
function Ve(e, t, n) {
  const [r, ...s] = T(t, `The ${e} component is already mounting`), [o, a] = /* @__PURE__ */ y(!1);
  return [
    (...i) => o() ? j.resolve() : r(...i).then((u) => {
      Rt(() => {
        o.set(!0), n(u);
      });
    }),
    ...s,
    [o, a]
  ];
}
const [bn, su] = /* @__PURE__ */ y({
  available: !1,
  type: "",
  accessGranted: !1,
  accessRequested: !1,
  deviceId: "",
  tokenSaved: !1
}), ou = /* @__PURE__ */ h(() => bn().available), kr = "web_app_biometry_get_info", au = /* @__PURE__ */ A(
  "requestBiometry",
  (e) => E(kr, "biometry_info_received", e),
  { isSupported: kr }
);
function Bs(e) {
  if (!eu(e))
    throw e;
}
function Ce(e) {
  const t = e();
  t && t.catch(Bs).cancel();
}
const Le = "biometry", st = "web_app_biometry_request_auth", yn = "biometry_info_received", js = (e) => {
  ot(_n(e));
};
function qs() {
  throw new ks("Biometry is not available");
}
function _n(e) {
  let t = !1, n = !1, r = "", s = !1, o = "", a = !1;
  return e.available && (t = !0, n = e.token_saved, r = e.device_id, s = e.access_requested, o = e.type, a = e.access_granted), { available: t, tokenSaved: n, deviceId: r, type: o, accessGranted: a, accessRequested: s };
}
const iu = G(st), [
  uu,
  cu,
  lu,
  Ms
] = /* @__PURE__ */ Ve(
  Le,
  (e) => {
    const t = z() && U(Le);
    return t ? j.resolve(t) : au({ abortSignal: e }).then(_n);
  },
  (e) => {
    C(yn, js), ot(e);
  }
), Ts = R(Le, st), gn = ue(Le, Ms[0], st), pu = Ts("mount", uu), [, $s, du] = cu, [, fu] = lu, [mu, hu] = Ms, [
  bu,
  yu,
  _u
] = T(
  (e) => j.fn(async (t) => {
    const n = bn();
    n.available || qs();
    const r = await E(st, "biometry_auth_requested", {
      ...e,
      ...t,
      params: { reason: ((e || {}).reason || "").trim() }
    }), { token: s } = r;
    return typeof s == "string" && ot({ ...n, token: s }), r;
  }, e),
  "Biometry authentication is already in progress"
), gu = gn("authenticate", bu), [, Is, vu] = yu, [, wu] = _u, Su = Ts("openSettings", () => {
  b("web_app_biometry_open_settings");
}), [
  ku,
  Eu,
  Au
] = T(
  (e) => j.fn(async (t) => {
    const n = await E("web_app_biometry_request_access", yn, {
      ...e,
      ...t,
      params: { reason: (e || {}).reason || "" }
    }).then(_n);
    return n.available || qs(), ot(n), n.accessGranted;
  }, e),
  "Biometry access request is already in progress"
), Cu = gn("requestAccess", ku), [, Os, xu] = Eu, [, Pu] = Au;
function ot(e) {
  bn.set(e), H(Le, e);
}
function Bu() {
  [Is, Os, $s].forEach(Ce), M(yn, js), mu.set(!1);
}
const ju = gn(
  "updateToken",
  (e) => (e || (e = {}), E("web_app_biometry_update_token", "biometry_token_updated", {
    ...e,
    params: {
      token: e.token || "",
      reason: e.reason
    }
  }).then((t) => t.status))
), Mf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  authError: wu,
  authPromise: Is,
  authenticate: gu,
  isAuthenticating: vu,
  isAvailable: ou,
  isMounted: hu,
  isMounting: du,
  isRequestingAccess: xu,
  isSupported: iu,
  mount: pu,
  mountError: fu,
  mountPromise: $s,
  openSettings: Su,
  requestAccess: Cu,
  requestAccessError: Pu,
  requestAccessPromise: Os,
  state: su,
  unmount: Bu,
  updateToken: ju
}, Symbol.toStringTag, { value: "Module" }));
function at(e, t) {
  return tt(e, { isMounted: t });
}
const ce = tt, it = "closingBehavior", [Er, qu] = /* @__PURE__ */ y(!1), [Ht, Ls] = /* @__PURE__ */ y(!1), Ds = at(it, Ls), Mu = ce(it), Tu = Ds("disableConfirmation", () => {
  vn(!1);
}), $u = Ds("enableConfirmation", () => {
  vn(!0);
}), Iu = Mu("mount", () => {
  Ht() || (vn(
    z() && U(it) || !1
  ), Ht.set(!0));
});
function vn(e) {
  e !== Er() && (b("web_app_setup_closing_behavior", { need_confirmation: e }), H(it, e), Er.set(e));
}
function Ou() {
  Ht.set(!1);
}
const Tf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  disableConfirmation: Tu,
  enableConfirmation: $u,
  isConfirmationEnabled: qu,
  isMounted: Ls,
  mount: Iu,
  unmount: Ou
}, Symbol.toStringTag, { value: "Module" }));
// @__NO_SIDE_EFFECTS__
function Rs(e) {
  return {
    lang: (e == null ? void 0 : e.lang) ?? void 0,
    message: e == null ? void 0 : e.message,
    abortEarly: (e == null ? void 0 : e.abortEarly) ?? void 0,
    abortPipeEarly: (e == null ? void 0 : e.abortPipeEarly) ?? void 0
  };
}
// @__NO_SIDE_EFFECTS__
function Lu(e, t) {
  var n;
  return (n = void 0) == null ? void 0 : n.get(t);
}
// @__NO_SIDE_EFFECTS__
function Du(e) {
  var t, n;
  const r = typeof e;
  return r === "string" ? `"${e}"` : r === "number" || r === "bigint" || r === "boolean" ? `${e}` : r === "object" || r === "function" ? (e && ((n = (t = Object.getPrototypeOf(e)) == null ? void 0 : t.constructor) == null ? void 0 : n.name)) ?? "null" : r;
}
function L(e, t, n, r, s) {
  const o = s && "input" in s ? s.input : n.value, a = (s == null ? void 0 : s.expected) ?? e.expects ?? null, i = (s == null ? void 0 : s.received) ?? /* @__PURE__ */ Du(o), u = {
    kind: e.kind,
    type: e.type,
    input: o,
    expected: a,
    received: i,
    message: `Invalid ${t}: ${a ? `Expected ${a} but r` : "R"}eceived ${i}`,
    requirement: e.requirement,
    path: s == null ? void 0 : s.path,
    issues: s == null ? void 0 : s.issues,
    lang: r.lang,
    abortEarly: r.abortEarly,
    abortPipeEarly: r.abortPipeEarly
  }, c = e.kind === "schema", l = (s == null ? void 0 : s.message) ?? e.message ?? /* @__PURE__ */ Lu(e.reference, u.lang) ?? (c ? (u.lang, void 0) : null) ?? r.message ?? (u.lang, void 0);
  l && (u.message = typeof l == "function" ? (
    // @ts-expect-error
    l(u)
  ) : l), c && (n.typed = !1), n.issues ? n.issues.push(u) : n.issues = [u];
}
// @__NO_SIDE_EFFECTS__
function Y(e) {
  return {
    version: 1,
    vendor: "valibot",
    validate(t) {
      return e["~run"]({ value: t }, /* @__PURE__ */ Rs());
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Vs(e, t) {
  return Object.hasOwn(e, t) && t !== "__proto__" && t !== "prototype" && t !== "constructor";
}
// @__NO_SIDE_EFFECTS__
function Ru(e, t) {
  const n = [...new Set(e)];
  return n.length > 1 ? `(${n.join(` ${t} `)})` : n[0] ?? "never";
}
var Ut = class extends Error {
  /**
   * Creates a Valibot error with useful information.
   *
   * @param issues The error issues.
   */
  constructor(e) {
    super(e[0].message), this.name = "ValiError", this.issues = e;
  }
};
// @__NO_SIDE_EFFECTS__
function Fs(e, t) {
  return {
    kind: "validation",
    type: "check",
    reference: Fs,
    async: !1,
    expects: null,
    requirement: e,
    message: t,
    "~run"(n, r) {
      return n.typed && !this.requirement(n.value) && L(this, "input", n, r), n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Ws(e) {
  return {
    kind: "validation",
    type: "integer",
    reference: Ws,
    async: !1,
    expects: null,
    requirement: Number.isInteger,
    message: e,
    "~run"(t, n) {
      return t.typed && !this.requirement(t.value) && L(this, "integer", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function wn(e) {
  return {
    kind: "transformation",
    type: "transform",
    reference: wn,
    async: !1,
    operation: e,
    "~run"(t) {
      return t.value = this.operation(t.value), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Ns(e, t, n) {
  return typeof e.default == "function" ? (
    // @ts-expect-error
    e.default(t, n)
  ) : (
    // @ts-expect-error
    e.default
  );
}
// @__NO_SIDE_EFFECTS__
function Hs(e, t) {
  return {
    kind: "schema",
    type: "array",
    reference: Hs,
    expects: "Array",
    async: !1,
    item: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ Y(this);
    },
    "~run"(n, r) {
      var s;
      const o = n.value;
      if (Array.isArray(o)) {
        n.typed = !0, n.value = [];
        for (let a = 0; a < o.length; a++) {
          const i = o[a], u = this.item["~run"]({ value: i }, r);
          if (u.issues) {
            const c = {
              type: "array",
              origin: "value",
              input: o,
              key: a,
              value: i
            };
            for (const l of u.issues)
              l.path ? l.path.unshift(c) : l.path = [c], (s = n.issues) == null || s.push(l);
            if (n.issues || (n.issues = u.issues), r.abortEarly) {
              n.typed = !1;
              break;
            }
          }
          u.typed || (n.typed = !1), n.value.push(u.value);
        }
      } else
        L(this, "type", n, r);
      return n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Sn(e) {
  return {
    kind: "schema",
    type: "date",
    reference: Sn,
    expects: "Date",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ Y(this);
    },
    "~run"(t, n) {
      return t.value instanceof Date ? isNaN(t.value) ? L(this, "type", t, n, {
        received: '"Invalid Date"'
      }) : t.typed = !0 : L(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Us(e, t) {
  return {
    kind: "schema",
    type: "instance",
    reference: Us,
    expects: e.name,
    async: !1,
    class: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ Y(this);
    },
    "~run"(n, r) {
      return n.value instanceof this.class ? n.typed = !0 : L(this, "type", n, r), n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function zt(e, t) {
  return {
    kind: "schema",
    type: "loose_object",
    reference: zt,
    expects: "Object",
    async: !1,
    entries: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ Y(this);
    },
    "~run"(n, r) {
      var s;
      const o = n.value;
      if (o && typeof o == "object") {
        n.typed = !0, n.value = {};
        for (const a in this.entries) {
          const i = this.entries[a];
          if (a in o || (i.type === "exact_optional" || i.type === "optional" || i.type === "nullish") && // @ts-expect-error
          i.default !== void 0) {
            const u = a in o ? (
              // @ts-expect-error
              o[a]
            ) : /* @__PURE__ */ Ns(i), c = i["~run"]({ value: u }, r);
            if (c.issues) {
              const l = {
                type: "object",
                origin: "value",
                input: o,
                key: a,
                value: u
              };
              for (const p of c.issues)
                p.path ? p.path.unshift(l) : p.path = [l], (s = n.issues) == null || s.push(p);
              if (n.issues || (n.issues = c.issues), r.abortEarly) {
                n.typed = !1;
                break;
              }
            }
            c.typed || (n.typed = !1), n.value[a] = c.value;
          } else if (i.type !== "exact_optional" && i.type !== "optional" && i.type !== "nullish" && (L(this, "key", n, r, {
            input: void 0,
            expected: `"${a}"`,
            path: [
              {
                type: "object",
                origin: "key",
                input: o,
                key: a,
                // @ts-expect-error
                value: o[a]
              }
            ]
          }), r.abortEarly))
            break;
        }
        if (!n.issues || !r.abortEarly)
          for (const a in o)
            /* @__PURE__ */ Vs(o, a) && !(a in this.entries) && (n.value[a] = o[a]);
      } else
        L(this, "type", n, r);
      return n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function kn(e) {
  return {
    kind: "schema",
    type: "number",
    reference: kn,
    expects: "number",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ Y(this);
    },
    "~run"(t, n) {
      return typeof t.value == "number" && !isNaN(t.value) ? t.typed = !0 : L(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function zs(e, t) {
  return {
    kind: "schema",
    type: "optional",
    reference: zs,
    expects: `(${e.expects} | undefined)`,
    async: !1,
    wrapped: e,
    default: t,
    get "~standard"() {
      return /* @__PURE__ */ Y(this);
    },
    "~run"(n, r) {
      return n.value === void 0 && (this.default !== void 0 && (n.value = /* @__PURE__ */ Ns(this, n, r)), n.value === void 0) ? (n.typed = !0, n) : this.wrapped["~run"](n, r);
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Gs(e, t, n) {
  return {
    kind: "schema",
    type: "record",
    reference: Gs,
    expects: "Object",
    async: !1,
    key: e,
    value: t,
    message: n,
    get "~standard"() {
      return /* @__PURE__ */ Y(this);
    },
    "~run"(r, s) {
      var o, a;
      const i = r.value;
      if (i && typeof i == "object") {
        r.typed = !0, r.value = {};
        for (const u in i)
          if (/* @__PURE__ */ Vs(i, u)) {
            const c = i[u], l = this.key["~run"]({ value: u }, s);
            if (l.issues) {
              const m = {
                type: "object",
                origin: "key",
                input: i,
                key: u,
                value: c
              };
              for (const f of l.issues)
                f.path = [m], (o = r.issues) == null || o.push(f);
              if (r.issues || (r.issues = l.issues), s.abortEarly) {
                r.typed = !1;
                break;
              }
            }
            const p = this.value["~run"](
              { value: c },
              s
            );
            if (p.issues) {
              const m = {
                type: "object",
                origin: "value",
                input: i,
                key: u,
                value: c
              };
              for (const f of p.issues)
                f.path ? f.path.unshift(m) : f.path = [m], (a = r.issues) == null || a.push(f);
              if (r.issues || (r.issues = p.issues), s.abortEarly) {
                r.typed = !1;
                break;
              }
            }
            (!l.typed || !p.typed) && (r.typed = !1), l.typed && (r.value[l.value] = p.value);
          }
      } else
        L(this, "type", r, s);
      return r;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function W(e) {
  return {
    kind: "schema",
    type: "string",
    reference: W,
    expects: "string",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ Y(this);
    },
    "~run"(t, n) {
      return typeof t.value == "string" ? t.typed = !0 : L(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Ar(e) {
  let t;
  if (e)
    for (const n of e)
      t ? t.push(...n.issues) : t = n.issues;
  return t;
}
// @__NO_SIDE_EFFECTS__
function Qs(e, t) {
  return {
    kind: "schema",
    type: "union",
    reference: Qs,
    expects: /* @__PURE__ */ Ru(
      e.map((n) => n.expects),
      "|"
    ),
    async: !1,
    options: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ Y(this);
    },
    "~run"(n, r) {
      let s, o, a;
      for (const i of this.options) {
        const u = i["~run"]({ value: n.value }, r);
        if (u.typed)
          if (u.issues)
            o ? o.push(u) : o = [u];
          else {
            s = u;
            break;
          }
        else
          a ? a.push(u) : a = [u];
      }
      if (s)
        return s;
      if (o) {
        if (o.length === 1)
          return o[0];
        L(this, "type", n, r, {
          issues: /* @__PURE__ */ Ar(o)
        }), n.typed = !0;
      } else {
        if ((a == null ? void 0 : a.length) === 1)
          return a[0];
        L(this, "type", n, r, {
          issues: /* @__PURE__ */ Ar(a)
        });
      }
      return n;
    }
  };
}
function ut(e, t, n) {
  const r = e["~run"]({ value: t }, /* @__PURE__ */ Rs(n));
  if (r.issues)
    throw new Ut(r.issues);
  return r.value;
}
// @__NO_SIDE_EFFECTS__
function Te(...e) {
  return {
    ...e[0],
    pipe: e,
    get "~standard"() {
      return /* @__PURE__ */ Y(this);
    },
    "~run"(t, n) {
      for (const r of e)
        if (r.kind !== "metadata") {
          if (t.issues && (r.kind === "schema" || r.kind === "transformation")) {
            t.typed = !1;
            break;
          }
          (!t.issues || !n.abortEarly && !n.abortPipeEarly) && (t = r["~run"](t, n));
        }
      return t;
    }
  };
}
const Js = "web_app_invoke_custom_method", Fe = R("cloudStorage", Js), Vu = G(Js), Ks = Fe("deleteItem", (e, t) => {
  const n = Array.isArray(e) ? e : [e];
  return n.length ? Ae("deleteStorageValues", { keys: n }, t).then() : j.resolve();
});
function Fu(e, t) {
  const n = Array.isArray(e) ? e : [e];
  return n.length ? Ae("getStorageValues", { keys: n }, t).then((r) => ut(
    /* @__PURE__ */ Gs(
      /* @__PURE__ */ Te(/* @__PURE__ */ W(), /* @__PURE__ */ Fs((s) => n.includes(s))),
      /* @__PURE__ */ W()
    ),
    r
  )) : j.resolve(typeof e == "string" ? "" : {});
}
const Wu = Fe("getItem", Fu), Zs = Fe("getKeys", (e) => Ae("getStorageKeys", {}, e).then(
  (t) => ut(/* @__PURE__ */ Hs(/* @__PURE__ */ W()), t)
)), Nu = Fe("setItem", (e, t, n) => Ae("saveStorageValue", {
  key: e,
  value: t
}, n).then()), Hu = Fe("clear", (e) => Zs(e).then(Ks)), $f = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  clear: Hu,
  deleteItem: Ks,
  getItem: Wu,
  getKeys: Zs,
  isSupported: Vu,
  setItem: Nu
}, Symbol.toStringTag, { value: "Module" })), We = "web_app_trigger_haptic_feedback", En = R("hapticFeedback", We), Uu = G(We), zu = En(
  "impactOccurred",
  (e) => {
    b(We, {
      type: "impact",
      impact_style: e
    });
  }
), Gu = En(
  "notificationOccurred",
  (e) => {
    b(We, {
      type: "notification",
      notification_type: e
    });
  }
), Qu = En(
  "selectionChanged",
  () => {
    b(We, { type: "selection_change" });
  }
), If = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  impactOccurred: zu,
  isSupported: Uu,
  notificationOccurred: Gu,
  selectionChanged: Qu
}, Symbol.toStringTag, { value: "Module" })), [Ys, Ju] = /* @__PURE__ */ y(void 0);
function X(e) {
  return /* @__PURE__ */ h(() => {
    const t = Ys();
    return t ? t[e] : void 0;
  });
}
const Xs = X("auth_date"), eo = X("can_send_after"), Ku = /* @__PURE__ */ h(() => {
  const e = Xs(), t = eo();
  return t && e ? new Date(e.getTime() + t * 1e3) : void 0;
}), Zu = X("chat"), Yu = X("chat_type"), Xu = X("chat_instance"), ec = X("hash"), tc = X("query_id"), [nc, rc] = /* @__PURE__ */ y(), sc = X("receiver");
function oc() {
  const e = ln();
  Ys.set(e.tgWebAppData), nc.set(Ai());
}
const ac = X("start_param"), ic = X("user"), Of = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  authDate: Xs,
  canSendAfter: eo,
  canSendAfterDate: Ku,
  chat: Zu,
  chatInstance: Xu,
  chatType: Yu,
  hash: ec,
  queryId: tc,
  raw: rc,
  receiver: sc,
  restore: oc,
  startParam: ac,
  state: Ju,
  user: ic
}, Symbol.toStringTag, { value: "Module" })), An = "web_app_open_invoice", uc = R("invoice", An), cc = G(An);
function lc(e, t, n) {
  let r;
  if (t === "url") {
    const { hostname: s, pathname: o } = new URL(e, window.location.href);
    if (s !== "t.me")
      throw new J(`Link has unexpected hostname: ${s}`);
    const a = o.match(/^\/(\$|invoice\/)([A-Za-z0-9\-_=]+)$/);
    if (!a)
      throw new J(
        'Expected to receive a link with a pathname in format "/invoice/{slug}" or "/${slug}"'
      );
    [, , r] = a;
  } else
    r = e, n = t;
  return E(An, "invoice_closed", {
    ...n,
    params: { slug: r },
    capture: (s) => r === s.slug
  }).then((s) => s.status);
}
const [
  pc,
  dc,
  fc
] = T(lc, "Invoice is already opened"), mc = uc("open", pc), [, hc, bc] = dc, [, yc] = fc, Lf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  isOpened: bc,
  isSupported: cc,
  open: mc,
  openError: yc,
  openPromise: hc
}, Symbol.toStringTag, { value: "Module" })), $e = "locationManager", Cn = "web_app_check_location", Cr = "web_app_open_location_settings", Ge = /* @__PURE__ */ Ee({
  available: !1,
  accessGranted: !1,
  accessRequested: !1
});
function xn(e) {
  return /* @__PURE__ */ h(() => Ge()[e]);
}
const _c = G(Cn), gc = xn("available"), vc = xn("accessGranted"), wc = xn("accessRequested");
function Sc(e) {
  let t = !1, n, r;
  return e.available && (t = !0, n = e.access_requested, r = e.access_granted), {
    available: t,
    accessGranted: r || !1,
    accessRequested: n || !1
  };
}
const [
  kc,
  Ec,
  Ac,
  to
] = /* @__PURE__ */ Ve(
  $e,
  (e) => {
    const t = z() && U($e);
    return t ? j.resolve(t) : E("web_app_check_location", "location_checked", e).then(Sc);
  },
  (e) => {
    Ge.set(e), H($e, e);
  }
), no = R($e, Cn), Cc = ue($e, to[0], Cn), xc = no("mount", kc), [, Pc, Bc] = Ec, [, jc] = Ac, [qc, Mc] = to, [
  Tc,
  $c,
  Ic
] = T(
  (e) => E("web_app_request_location", "location_requested", e).then((t) => {
    if (!t.available)
      throw Ge.set({ ...Ge(), available: !1 }), new ks("Location data tracking is not available");
    const { available: n, ...r } = t;
    return r;
  }),
  "Location request is currently in progress"
), Oc = Cc("requestLocation", Tc), [, ro, Lc] = $c, [, Dc] = Ic, Rc = no("openSettings", () => {
  b(Cr);
}, Cr);
function Vc() {
  Ce(ro), qc.set(!1);
}
const Df = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  isAccessGranted: vc,
  isAccessRequested: wc,
  isAvailable: gc,
  isMounted: Mc,
  isMounting: Bc,
  isRequestingLocation: Lc,
  isSupported: _c,
  mount: xc,
  mountError: jc,
  mountPromise: Pc,
  openSettings: Rc,
  requestLocation: Oc,
  requestLocationError: Dc,
  requestLocationPromise: ro,
  unmount: Vc
}, Symbol.toStringTag, { value: "Module" }));
function Pn(e) {
  const t = {};
  for (const n in e) {
    const r = e[n];
    r !== void 0 && (t[n] = r);
  }
  return t;
}
// @__NO_SIDE_EFFECTS__
function so(e) {
  return {
    lang: (e == null ? void 0 : e.lang) ?? void 0,
    message: e == null ? void 0 : e.message,
    abortEarly: (e == null ? void 0 : e.abortEarly) ?? void 0,
    abortPipeEarly: (e == null ? void 0 : e.abortPipeEarly) ?? void 0
  };
}
// @__NO_SIDE_EFFECTS__
function Fc(e, t) {
  var n;
  return (n = void 0) == null ? void 0 : n.get(t);
}
// @__NO_SIDE_EFFECTS__
function Wc(e) {
  var t, n;
  const r = typeof e;
  return r === "string" ? `"${e}"` : r === "number" || r === "bigint" || r === "boolean" ? `${e}` : r === "object" || r === "function" ? (e && ((n = (t = Object.getPrototypeOf(e)) == null ? void 0 : t.constructor) == null ? void 0 : n.name)) ?? "null" : r;
}
function D(e, t, n, r, s) {
  const o = s && "input" in s ? s.input : n.value, a = (s == null ? void 0 : s.expected) ?? e.expects ?? null, i = (s == null ? void 0 : s.received) ?? /* @__PURE__ */ Wc(o), u = {
    kind: e.kind,
    type: e.type,
    input: o,
    expected: a,
    received: i,
    message: `Invalid ${t}: ${a ? `Expected ${a} but r` : "R"}eceived ${i}`,
    requirement: e.requirement,
    path: s == null ? void 0 : s.path,
    issues: s == null ? void 0 : s.issues,
    lang: r.lang,
    abortEarly: r.abortEarly,
    abortPipeEarly: r.abortPipeEarly
  }, c = e.kind === "schema", l = (s == null ? void 0 : s.message) ?? e.message ?? /* @__PURE__ */ Fc(e.reference, u.lang) ?? (c ? (u.lang, void 0) : null) ?? r.message ?? (u.lang, void 0);
  l && (u.message = typeof l == "function" ? (
    // @ts-expect-error
    l(u)
  ) : l), c && (n.typed = !1), n.issues ? n.issues.push(u) : n.issues = [u];
}
// @__NO_SIDE_EFFECTS__
function Q(e) {
  return {
    version: 1,
    vendor: "valibot",
    validate(t) {
      return e["~run"]({ value: t }, /* @__PURE__ */ so());
    }
  };
}
// @__NO_SIDE_EFFECTS__
function oo(e, t) {
  return Object.hasOwn(e, t) && t !== "__proto__" && t !== "prototype" && t !== "constructor";
}
// @__NO_SIDE_EFFECTS__
function Nc(e, t) {
  const n = [...new Set(e)];
  return n.length > 1 ? `(${n.join(` ${t} `)})` : n[0] ?? "never";
}
var Hc = class extends Error {
  /**
   * Creates a Valibot error with useful information.
   *
   * @param issues The error issues.
   */
  constructor(e) {
    super(e[0].message), this.name = "ValiError", this.issues = e;
  }
};
// @__NO_SIDE_EFFECTS__
function ao(e, t) {
  return {
    kind: "validation",
    type: "check",
    reference: ao,
    async: !1,
    expects: null,
    requirement: e,
    message: t,
    "~run"(n, r) {
      return n.typed && !this.requirement(n.value) && D(this, "input", n, r), n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function io(e) {
  return {
    kind: "validation",
    type: "integer",
    reference: io,
    async: !1,
    expects: null,
    requirement: Number.isInteger,
    message: e,
    "~run"(t, n) {
      return t.typed && !this.requirement(t.value) && D(this, "integer", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function ae(e) {
  return {
    kind: "transformation",
    type: "transform",
    reference: ae,
    async: !1,
    operation: e,
    "~run"(t) {
      return t.value = this.operation(t.value), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function uo(e, t, n) {
  return typeof e.default == "function" ? (
    // @ts-expect-error
    e.default(t, n)
  ) : (
    // @ts-expect-error
    e.default
  );
}
// @__NO_SIDE_EFFECTS__
function qe(e) {
  return {
    kind: "schema",
    type: "boolean",
    reference: qe,
    expects: "boolean",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(t, n) {
      return typeof t.value == "boolean" ? t.typed = !0 : D(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function co(e) {
  return {
    kind: "schema",
    type: "date",
    reference: co,
    expects: "Date",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(t, n) {
      return t.value instanceof Date ? isNaN(t.value) ? D(this, "type", t, n, {
        received: '"Invalid Date"'
      }) : t.typed = !0 : D(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function lo(e, t) {
  return {
    kind: "schema",
    type: "instance",
    reference: lo,
    expects: e.name,
    async: !1,
    class: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(n, r) {
      return n.value instanceof this.class ? n.typed = !0 : D(this, "type", n, r), n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Bn(e) {
  return {
    kind: "schema",
    type: "lazy",
    reference: Bn,
    expects: "unknown",
    async: !1,
    getter: e,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(t, n) {
      return this.getter(t.value)["~run"](t, n);
    }
  };
}
// @__NO_SIDE_EFFECTS__
function Ne(e, t) {
  return {
    kind: "schema",
    type: "loose_object",
    reference: Ne,
    expects: "Object",
    async: !1,
    entries: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(n, r) {
      var s;
      const o = n.value;
      if (o && typeof o == "object") {
        n.typed = !0, n.value = {};
        for (const a in this.entries) {
          const i = this.entries[a];
          if (a in o || (i.type === "exact_optional" || i.type === "optional" || i.type === "nullish") && // @ts-expect-error
          i.default !== void 0) {
            const u = a in o ? (
              // @ts-expect-error
              o[a]
            ) : /* @__PURE__ */ uo(i), c = i["~run"]({ value: u }, r);
            if (c.issues) {
              const l = {
                type: "object",
                origin: "value",
                input: o,
                key: a,
                value: u
              };
              for (const p of c.issues)
                p.path ? p.path.unshift(l) : p.path = [l], (s = n.issues) == null || s.push(p);
              if (n.issues || (n.issues = c.issues), r.abortEarly) {
                n.typed = !1;
                break;
              }
            }
            c.typed || (n.typed = !1), n.value[a] = c.value;
          } else if (i.type !== "exact_optional" && i.type !== "optional" && i.type !== "nullish" && (D(this, "key", n, r, {
            input: void 0,
            expected: `"${a}"`,
            path: [
              {
                type: "object",
                origin: "key",
                input: o,
                key: a,
                // @ts-expect-error
                value: o[a]
              }
            ]
          }), r.abortEarly))
            break;
        }
        if (!n.issues || !r.abortEarly)
          for (const a in o)
            /* @__PURE__ */ oo(o, a) && !(a in this.entries) && (n.value[a] = o[a]);
      } else
        D(this, "type", n, r);
      return n;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function ct(e) {
  return {
    kind: "schema",
    type: "number",
    reference: ct,
    expects: "number",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(t, n) {
      return typeof t.value == "number" && !isNaN(t.value) ? t.typed = !0 : D(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function w(e, t) {
  return {
    kind: "schema",
    type: "optional",
    reference: w,
    expects: `(${e.expects} | undefined)`,
    async: !1,
    wrapped: e,
    default: t,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(n, r) {
      return n.value === void 0 && (this.default !== void 0 && (n.value = /* @__PURE__ */ uo(this, n, r)), n.value === void 0) ? (n.typed = !0, n) : this.wrapped["~run"](n, r);
    }
  };
}
// @__NO_SIDE_EFFECTS__
function po(e, t, n) {
  return {
    kind: "schema",
    type: "record",
    reference: po,
    expects: "Object",
    async: !1,
    key: e,
    value: t,
    message: n,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(r, s) {
      var o, a;
      const i = r.value;
      if (i && typeof i == "object") {
        r.typed = !0, r.value = {};
        for (const u in i)
          if (/* @__PURE__ */ oo(i, u)) {
            const c = i[u], l = this.key["~run"]({ value: u }, s);
            if (l.issues) {
              const m = {
                type: "object",
                origin: "key",
                input: i,
                key: u,
                value: c
              };
              for (const f of l.issues)
                f.path = [m], (o = r.issues) == null || o.push(f);
              if (r.issues || (r.issues = l.issues), s.abortEarly) {
                r.typed = !1;
                break;
              }
            }
            const p = this.value["~run"](
              { value: c },
              s
            );
            if (p.issues) {
              const m = {
                type: "object",
                origin: "value",
                input: i,
                key: u,
                value: c
              };
              for (const f of p.issues)
                f.path ? f.path.unshift(m) : f.path = [m], (a = r.issues) == null || a.push(f);
              if (r.issues || (r.issues = p.issues), s.abortEarly) {
                r.typed = !1;
                break;
              }
            }
            (!l.typed || !p.typed) && (r.typed = !1), l.typed && (r.value[l.value] = p.value);
          }
      } else
        D(this, "type", r, s);
      return r;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function g(e) {
  return {
    kind: "schema",
    type: "string",
    reference: g,
    expects: "string",
    async: !1,
    message: e,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(t, n) {
      return typeof t.value == "string" ? t.typed = !0 : D(this, "type", t, n), t;
    }
  };
}
// @__NO_SIDE_EFFECTS__
function xr(e) {
  let t;
  if (e)
    for (const n of e)
      t ? t.push(...n.issues) : t = n.issues;
  return t;
}
// @__NO_SIDE_EFFECTS__
function jn(e, t) {
  return {
    kind: "schema",
    type: "union",
    reference: jn,
    expects: /* @__PURE__ */ Nc(
      e.map((n) => n.expects),
      "|"
    ),
    async: !1,
    options: e,
    message: t,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(n, r) {
      let s, o, a;
      for (const i of this.options) {
        const u = i["~run"]({ value: n.value }, r);
        if (u.typed)
          if (u.issues)
            o ? o.push(u) : o = [u];
          else {
            s = u;
            break;
          }
        else
          a ? a.push(u) : a = [u];
      }
      if (s)
        return s;
      if (o) {
        if (o.length === 1)
          return o[0];
        D(this, "type", n, r, {
          issues: /* @__PURE__ */ xr(o)
        }), n.typed = !0;
      } else {
        if ((a == null ? void 0 : a.length) === 1)
          return a[0];
        D(this, "type", n, r, {
          issues: /* @__PURE__ */ xr(a)
        });
      }
      return n;
    }
  };
}
function fo(e, t, n) {
  const r = e["~run"]({ value: t }, /* @__PURE__ */ so(n));
  if (r.issues)
    throw new Hc(r.issues);
  return r.value;
}
// @__NO_SIDE_EFFECTS__
function ie(...e) {
  return {
    ...e[0],
    pipe: e,
    get "~standard"() {
      return /* @__PURE__ */ Q(this);
    },
    "~run"(t, n) {
      for (const r of e)
        if (r.kind !== "metadata") {
          if (t.issues && (r.kind === "schema" || r.kind === "transformation")) {
            t.typed = !1;
            break;
          }
          (!t.issues || !n.abortEarly && !n.abortPipeEarly) && (t = r["~run"](t, n));
        }
      return t;
    }
  };
}
function Uc(e) {
  return e.replace(/_[a-z]/g, (t) => t[1].toUpperCase());
}
function zc(e) {
  return Object.entries(e).reduce((t, [n, r]) => (t[Uc(n)] = r, t), {});
}
function Gt(e) {
  const t = zc(e);
  for (const n in t) {
    const r = t[n];
    r && typeof r == "object" && !(r instanceof Date) && (t[n] = Array.isArray(r) ? r.map(Gt) : Gt(r));
  }
  return t;
}
function qn(e) {
  return /* @__PURE__ */ ae((t) => e ? Gt(t) : t);
}
function mo(e) {
  return (t) => /* @__PURE__ */ ie(
    e,
    qn(t)
  );
}
function ho(e) {
  return (t, n) => fo(
    /* @__PURE__ */ ie(e, qn(n)),
    t
  );
}
function bo() {
  return /* @__PURE__ */ ae(JSON.parse);
}
function Mn(e) {
  const t = mo(e);
  return (n) => /* @__PURE__ */ ie(
    /* @__PURE__ */ g(),
    bo(),
    t(n)
  );
}
function yo(e) {
  return /* @__PURE__ */ ae((t) => {
    const n = {};
    return new URLSearchParams(t).forEach((r, s) => {
      const o = n[s];
      Array.isArray(o) ? o.push(r) : o === void 0 ? n[s] = r : n[s] = [o, r];
    }), fo(e, n);
  });
}
function _o(e) {
  return (t) => /* @__PURE__ */ ie(
    /* @__PURE__ */ jn([/* @__PURE__ */ g(), /* @__PURE__ */ lo(URLSearchParams)]),
    yo(e),
    qn(t)
  );
}
const Pr = /* @__PURE__ */ w(/* @__PURE__ */ Bn(() => Zc())), Gc = /* @__PURE__ */ Ne({
  id: /* @__PURE__ */ ct(),
  photo_url: /* @__PURE__ */ w(/* @__PURE__ */ g()),
  type: /* @__PURE__ */ g(),
  title: /* @__PURE__ */ g(),
  username: /* @__PURE__ */ w(/* @__PURE__ */ g())
}), Qc = /* @__PURE__ */ Ne({
  added_to_attachment_menu: /* @__PURE__ */ w(/* @__PURE__ */ qe()),
  allows_write_to_pm: /* @__PURE__ */ w(/* @__PURE__ */ qe()),
  first_name: /* @__PURE__ */ g(),
  id: /* @__PURE__ */ ct(),
  is_bot: /* @__PURE__ */ w(/* @__PURE__ */ qe()),
  is_premium: /* @__PURE__ */ w(/* @__PURE__ */ qe()),
  last_name: /* @__PURE__ */ w(/* @__PURE__ */ g()),
  language_code: /* @__PURE__ */ w(/* @__PURE__ */ g()),
  photo_url: /* @__PURE__ */ w(/* @__PURE__ */ g()),
  username: /* @__PURE__ */ w(/* @__PURE__ */ g())
}), Jc = /* @__PURE__ */ Ne({
  auth_date: /* @__PURE__ */ ie(
    /* @__PURE__ */ g(),
    /* @__PURE__ */ ae((e) => new Date(Number(e) * 1e3)),
    /* @__PURE__ */ co()
  ),
  can_send_after: /* @__PURE__ */ w(/* @__PURE__ */ ie(/* @__PURE__ */ g(), /* @__PURE__ */ ae(Number), /* @__PURE__ */ io())),
  chat: /* @__PURE__ */ w(/* @__PURE__ */ Bn(() => Kc())),
  chat_type: /* @__PURE__ */ w(/* @__PURE__ */ g()),
  chat_instance: /* @__PURE__ */ w(/* @__PURE__ */ g()),
  hash: /* @__PURE__ */ g(),
  query_id: /* @__PURE__ */ w(/* @__PURE__ */ g()),
  receiver: Pr,
  start_param: /* @__PURE__ */ w(/* @__PURE__ */ g()),
  signature: /* @__PURE__ */ g(),
  user: Pr
}), Kc = Mn(Gc), Zc = Mn(Qc), go = _o(Jc);
function ve(e) {
  return /^#[\da-f]{6}$/i.test(e);
}
function Yc(e) {
  return /^#[\da-f]{3}$/i.test(e);
}
function Xc(e) {
  const t = e.replace(/\s/g, "").toLowerCase();
  if (ve(t))
    return t;
  if (Yc(t)) {
    let r = "#";
    for (let s = 0; s < 3; s += 1)
      r += t[1 + s].repeat(2);
    return r;
  }
  const n = t.match(/^rgb\((\d{1,3}),(\d{1,3}),(\d{1,3})\)$/) || t.match(/^rgba\((\d{1,3}),(\d{1,3}),(\d{1,3}),\d{1,3}\)$/);
  if (!n)
    throw new Error(`Value "${e}" does not satisfy any of known RGB formats.`);
  return n.slice(1).reduce((r, s) => r + parseInt(s, 10).toString(16).padStart(2, "0"), "#");
}
const el = mo(
  /* @__PURE__ */ po(
    /* @__PURE__ */ g(),
    /* @__PURE__ */ ie(
      /* @__PURE__ */ jn([/* @__PURE__ */ g(), /* @__PURE__ */ ct()]),
      /* @__PURE__ */ ae((e) => typeof e == "number" ? `#${(e & 16777215).toString(16).padStart(6, "0")}` : e),
      /* @__PURE__ */ ao(ve)
    )
  )
), Ct = /* @__PURE__ */ w(
  /* @__PURE__ */ ie(/* @__PURE__ */ g(), /* @__PURE__ */ ae((e) => e === "1"))
), Br = Mn(el()), tl = /* @__PURE__ */ Ne({
  tgWebAppBotInline: Ct,
  tgWebAppData: /* @__PURE__ */ w(go()),
  tgWebAppDefaultColors: /* @__PURE__ */ w(Br()),
  tgWebAppFullscreen: Ct,
  tgWebAppPlatform: /* @__PURE__ */ g(),
  tgWebAppShowSettings: Ct,
  tgWebAppStartParam: /* @__PURE__ */ w(/* @__PURE__ */ g()),
  tgWebAppThemeParams: Br(),
  tgWebAppVersion: /* @__PURE__ */ g()
}), nl = _o(tl), Rf = ho(go()), Vf = ho(nl());
function vo(e, t) {
  return t || (t = (n, r) => JSON.stringify(r)), new URLSearchParams(
    Object.entries(e).reduce((n, [r, s]) => (Array.isArray(s) ? n.push(...s.map((o) => [r, String(o)])) : s != null && n.push([
      r,
      s instanceof Date ? (s.getTime() / 1e3 | 0).toString() : typeof s == "string" || typeof s == "number" ? String(s) : typeof s == "boolean" ? s ? "1" : "0" : t(r, s)
    ]), n), [])
  ).toString();
}
function rl(e) {
  return vo(e);
}
function Ff(e) {
  return vo(e, (t, n) => t === "tgWebAppData" ? rl(n) : JSON.stringify(n));
}
function wo(e) {
  const t = Xc(e);
  return Math.sqrt(
    [0.299, 0.587, 0.114].reduce((n, r, s) => {
      const o = parseInt(t.slice(1 + s * 2, 1 + (s + 1) * 2), 16);
      return n + o * o * r;
    }, 0)
  ) < 120;
}
const [xt, sl] = /* @__PURE__ */ y(!1), [_e, ol] = /* @__PURE__ */ y({});
function q(e) {
  return /* @__PURE__ */ h(() => _e()[e]);
}
const al = q("accent_text_color"), lt = q("bg_color"), Tn = q("button_color"), So = q("button_text_color"), ko = q("bottom_bar_bg_color"), il = q("destructive_text_color"), ul = q("header_bg_color"), cl = q("hint_color"), ll = /* @__PURE__ */ h(() => {
  const e = lt();
  return !e || wo(e);
}), pl = q("link_color"), Qe = q("secondary_bg_color"), dl = q("section_bg_color"), fl = q("section_header_text_color"), ml = q("section_separator_color"), hl = q("subtitle_text_color"), bl = q("text_color");
function fe(e) {
  return /* @__PURE__ */ h(() => $n()[e]);
}
const Ie = /* @__PURE__ */ Ee({
  hasShineEffect: !1,
  isEnabled: !0,
  isLoaderVisible: !1,
  isVisible: !1,
  text: "Continue"
}), $n = /* @__PURE__ */ h(() => {
  const e = Ie();
  return {
    ...e,
    backgroundColor: e.backgroundColor || Tn() || "#2481cc",
    textColor: e.textColor || So() || "#ffffff"
  };
}), [Qt, Eo] = /* @__PURE__ */ y(!1), yl = fe("backgroundColor"), _l = fe("hasShineEffect"), gl = fe("isEnabled"), vl = fe("isLoaderVisible"), wl = fe("isVisible"), Sl = fe("text"), kl = fe("textColor"), El = "web_app_setup_main_button", Ao = "main_button_pressed", pt = "mainButton", In = ce(pt), Al = at(pt, Eo), Cl = In("mount", () => {
  if (!Qt()) {
    const e = z() && U(pt);
    e && Ie.set(e), Qt.set(!0);
  }
}), xl = In(
  "onClick",
  (e) => C(Ao, e)
), Pl = In(
  "offClick",
  (e) => {
    M(Ao, e);
  }
), Bl = Al(
  "setParams",
  (e) => {
    Ie.set({ ...Ie(), ...Pn(e) }), H(pt, Ie());
    const t = $n();
    t.text && b(El, {
      color: t.backgroundColor,
      has_shine_effect: t.hasShineEffect,
      is_active: t.isEnabled,
      is_progress_visible: t.isLoaderVisible,
      is_visible: t.isVisible,
      text: t.text,
      text_color: t.textColor
    });
  }
);
function jl() {
  Qt.set(!1);
}
const Wf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  backgroundColor: yl,
  hasShineEffect: _l,
  isEnabled: gl,
  isLoaderVisible: vl,
  isMounted: Eo,
  isVisible: wl,
  mount: Cl,
  offClick: Pl,
  onClick: xl,
  setParams: Bl,
  state: $n,
  text: Sl,
  textColor: kl,
  unmount: jl
}, Symbol.toStringTag, { value: "Module" }));
function On(e, t) {
  document.documentElement.style.setProperty(e, t);
}
function Ln(e) {
  document.documentElement.style.removeProperty(e);
}
const De = "themeParams", Co = "theme_changed", ql = ce(De), xo = ({ theme_params: e }) => {
  _e.set(e), H(De, e);
}, [
  Ml,
  Tl,
  $l,
  Po
] = /* @__PURE__ */ Ve(
  De,
  (e) => {
    const t = z() && U(De);
    return t ? j.resolve(t) : E("web_app_request_theme", "theme_changed", e).then((n) => n.theme_params);
  },
  (e) => {
    C(Co, xo), _e.set(e);
  }
), Il = at(De, Po[0]), Ol = Il(
  "bindCssVars",
  (e) => {
    if (xt())
      throw new fn();
    e || (e = (r) => `--tg-theme-${Bi(r)}`);
    function t(r) {
      Object.entries(_e()).forEach(([s, o]) => {
        o && r(s, o);
      });
    }
    function n() {
      t((r, s) => {
        On(e(r), s);
      });
    }
    return n(), _e.sub(n), xt.set(!0), () => {
      t(Ln), _e.unsub(n), xt.set(!1);
    };
  }
), Bo = ql("mount", Ml), [, jo, Ll] = Tl, [, Dl] = $l, [Rl, Vl] = Po;
function Fl() {
  Ce(jo), M(Co, xo), Rl.set(!1);
}
// @__NO_SIDE_EFFECTS__
function qo(e) {
  return /* @__PURE__ */ h(() => {
    const t = e();
    return ve(t) ? t : t === "bg_color" ? lt() : Qe();
  });
}
const [Je, Wl] = /* @__PURE__ */ y("bg_color"), Dn = /* @__PURE__ */ qo(Je), [Ke, Nl] = /* @__PURE__ */ y("bottom_bar_bg_color"), Rn = /* @__PURE__ */ h(() => {
  const e = Ke();
  return ve(e) ? e : e === "bottom_bar_bg_color" ? ko() || Qe() : e === "secondary_bg_color" ? Qe() : lt();
}), [Ze, Hl] = /* @__PURE__ */ y("bg_color"), Mo = /* @__PURE__ */ qo(Ze), [Pt, Ul] = /* @__PURE__ */ y(!1), zl = /* @__PURE__ */ h(() => {
  const e = Dn();
  return e ? wo(e) : !1;
}), [Vn, Gl] = /* @__PURE__ */ y(!0), To = /* @__PURE__ */ h(() => ({
  backgroundColor: Je(),
  bottomBarColor: Ke(),
  headerColor: Ze(),
  isActive: Vn()
})), Jt = "web_app_set_background_color", Kt = "web_app_set_bottom_bar_color", Ue = "web_app_set_header_color", $o = "visibility_changed", we = "miniApp", Fn = {
  any: [
    Jt,
    Kt,
    Ue
  ]
}, Ql = /* @__PURE__ */ h(() => Fn.any.some((e) => pe(e, re()))), Io = (e) => {
  Vn.set(e.is_visible), ft();
}, [
  Jl,
  Kl,
  Zl,
  Oo
] = /* @__PURE__ */ Ve(
  we,
  (e) => Bo(e).then(() => z() && U(we) || void 0),
  (e) => {
    Ro.ifAvailable(e ? e.backgroundColor : "bg_color"), Vo.ifAvailable(e ? e.bottomBarColor : "bottom_bar_bg_color"), Fo.ifAvailable(e ? e.headerColor : "bg_color"), Vn.set(e ? e.isActive : !0), C($o, Io);
  }
), Lo = ce(we), Yl = R(we, Fn), dt = ue(we, Oo[0], Fn), Xl = dt(
  "bindCssVars",
  (e) => {
    if (Pt())
      throw new fn();
    const [t, n] = pn();
    function r(s, o) {
      function a() {
        On(s, o() || null);
      }
      a(), t(o.sub(a), Ln.bind(null, s));
    }
    return e || (e = (s) => `--tg-${ys(s)}`), r(e("bgColor"), Dn), r(e("bottomBarColor"), Rn), r(e("headerColor"), Mo), t(() => {
      Pt.set(!1);
    }), Pt.set(!0), n;
  }
), ep = Lo("close", (e) => {
  b("web_app_close", { return_back: e });
}), tp = Yl("mount", Jl), [, Do, np] = Kl, [, rp] = Zl, [sp, op] = Oo, ap = Lo("ready", () => {
  b("web_app_ready");
});
function ft() {
  H(we, To());
}
const Ro = dt(
  "setBackgroundColor",
  (e) => {
    e !== Je() && (b(Jt, { color: e }), Je.set(e), ft());
  },
  Jt
), Vo = dt(
  "setBottomBarColor",
  (e) => {
    e !== Ke() && (b(Kt, { color: e }), Ke.set(e), ft());
  },
  Kt
), Fo = dt(
  "setHeaderColor",
  (e) => {
    e !== Ze() && (b(Ue, ve(e) ? { color: e } : { color_key: e }), Ze.set(e), ft());
  },
  Ue,
  {
    rgb: [Ue, "color", ve]
  }
);
function ip() {
  Ce(Do), M($o, Io), sp.set(!1);
}
const Nf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  backgroundColor: Wl,
  backgroundColorRGB: Dn,
  bindCssVars: Xl,
  bottomBarColor: Nl,
  bottomBarColorRGB: Rn,
  close: ep,
  headerColor: Hl,
  headerColorRGB: Mo,
  isActive: Gl,
  isCssVarsBound: Ul,
  isDark: zl,
  isMounted: op,
  isMounting: np,
  isSupported: Ql,
  mount: tp,
  mountError: rp,
  mountPromise: Do,
  ready: ap,
  setBackgroundColor: Ro,
  setBottomBarColor: Vo,
  setHeaderColor: Fo,
  state: To,
  unmount: ip
}, Symbol.toStringTag, { value: "Module" }));
function up(e) {
  const t = e.message.trim(), n = (e.title || "").trim(), r = e.buttons || [];
  if (n.length > 64)
    throw new J(`Invalid title: ${n}`);
  if (!t || t.length > 256)
    throw new J(`Invalid message: ${t}`);
  if (r.length > 3)
    throw new J(`Invalid buttons count: ${r.length}`);
  return {
    title: n,
    message: t,
    buttons: r.length ? r.map((s, o) => {
      const a = s.id || "";
      if (a.length > 64)
        throw new J(`Button with index ${o} has invalid id: ${a}`);
      if (!s.type || s.type === "default" || s.type === "destructive") {
        const i = s.text.trim();
        if (!i || i.length > 64)
          throw new J(`Button with index ${o} has invalid text: ${i}`);
        return { type: s.type, text: i, id: a };
      }
      return { type: s.type, id: a };
    }) : [{ type: "close", id: "" }]
  };
}
const Wn = "web_app_open_popup", cp = R("popup", Wn), lp = G(Wn), [
  pp,
  dp,
  fp
] = T((e) => E(Wn, "popup_closed", {
  ...e,
  params: up(e)
}).then(({ button_id: t }) => t === void 0 ? null : t), "A popup is already opened"), mp = cp("open", pp), [, hp, bp] = dp, [, yp] = fp, Hf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  isOpened: bp,
  isSupported: lp,
  open: mp,
  openError: yp,
  openPromise: hp
}, Symbol.toStringTag, { value: "Module" })), Wo = "web_app_close_scan_qr_popup", Nn = "web_app_open_scan_qr_popup", _p = "scan_qr_popup_closed", gp = "qr_text_received", No = R("qrScanner", Nn), vp = No("close", () => {
  b(Wo), Ce(Ho);
}), wp = G(Nn);
function Sp(e) {
  e || (e = {});
  const { onCaptured: t, text: n, capture: r } = e, [, s] = pn(
    C(_p, () => {
      o.resolve();
    }),
    C(gp, (a) => {
      t ? t(a.data) : (!r || r(a.data)) && (o.resolve(a.data), b(Wo));
    })
  ), o = new ru(e);
  return (e.postEvent || b)(Nn, { text: n }), j.resolve(o).catch(Bs).finally(s);
}
const [
  kp,
  Ep,
  Ap
] = T(Sp, "The QR Scanner is already opened"), Cp = No("open", kp), [, Ho, xp] = Ep, [, Pp] = Ap, Uf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  close: vp,
  isOpened: xp,
  isSupported: wp,
  open: Cp,
  openError: Pp,
  openPromise: Ho
}, Symbol.toStringTag, { value: "Module" }));
function le(e) {
  return /* @__PURE__ */ h(() => Hn()[e]);
}
const Oe = /* @__PURE__ */ Ee({
  hasShineEffect: !1,
  isEnabled: !0,
  isLoaderVisible: !1,
  isVisible: !1,
  position: "left",
  text: "Cancel"
}), Hn = /* @__PURE__ */ h(() => {
  const e = Oe();
  return {
    ...e,
    backgroundColor: e.backgroundColor || Rn() || "#000000",
    textColor: e.textColor || Tn() || "#2481cc"
  };
}), [Zt, Uo] = /* @__PURE__ */ y(!1), Bp = le("backgroundColor"), jp = le("hasShineEffect"), qp = le("isEnabled"), Mp = le("isLoaderVisible"), Tp = le("isVisible"), $p = le("position"), Ip = le("text"), Op = le("textColor"), mt = "web_app_setup_secondary_button", zo = "secondary_button_pressed", ht = "secondaryButton", Un = R(ht, mt), Lp = ue(ht, Uo, mt), Dp = G(mt), Rp = Un("mount", () => {
  if (!Zt()) {
    const e = z() && U(ht);
    e && Oe.set(e), Zt.set(!0);
  }
}), Vp = Un(
  "onClick",
  (e) => C(zo, e)
), Fp = Un(
  "offClick",
  (e) => {
    M(zo, e);
  }
), Wp = Lp(
  "setParams",
  (e) => {
    Oe.set({ ...Oe(), ...Pn(e) }), H(ht, Oe());
    const t = Hn();
    t.text && b(mt, {
      color: t.backgroundColor,
      has_shine_effect: t.hasShineEffect,
      is_active: t.isEnabled,
      is_progress_visible: t.isLoaderVisible,
      is_visible: t.isVisible,
      position: t.position,
      text: t.text,
      text_color: t.textColor
    });
  }
);
function Np() {
  Zt.set(!1);
}
const zf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  backgroundColor: Bp,
  hasShineEffect: jp,
  isEnabled: qp,
  isLoaderVisible: Mp,
  isMounted: Uo,
  isSupported: Dp,
  isVisible: Tp,
  mount: Rp,
  offClick: Fp,
  onClick: Vp,
  position: $p,
  setParams: Wp,
  state: Hn,
  text: Ip,
  textColor: Op,
  unmount: Np
}, Symbol.toStringTag, { value: "Module" })), bt = "web_app_setup_settings_button", Go = "settings_button_pressed", yt = "settingsButton", [jr, Hp] = /* @__PURE__ */ y(!1), [Ye, Up] = /* @__PURE__ */ y(!1), zp = G(bt), zn = R(yt, bt), Qo = ue(yt, Ye, bt), Gp = Qo("hide", () => {
  Gn(!1);
}), Qp = zn("mount", () => {
  Ye() || (Gn(z() && U(yt) || !1), Ye.set(!0));
});
function Gn(e) {
  e !== jr() && (b(bt, { is_visible: e }), H(yt, e), jr.set(e));
}
const Jp = zn(
  "onClick",
  (e) => C(Go, e)
), Kp = zn(
  "offClick",
  (e) => {
    M(Go, e);
  }
), Zp = Qo("show", () => {
  Gn(!0);
});
function Yp() {
  Ye.set(!1);
}
const Gf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  hide: Gp,
  isMounted: Up,
  isSupported: zp,
  isVisible: Hp,
  mount: Qp,
  offClick: Kp,
  onClick: Jp,
  show: Zp,
  unmount: Yp
}, Symbol.toStringTag, { value: "Module" })), _t = "web_app_setup_swipe_behavior", gt = "swipeBehavior", [Re, Xp] = /* @__PURE__ */ y(!1), ed = G(_t), [Yt, td] = /* @__PURE__ */ y(!0), nd = R(gt, _t), Jo = ue(gt, Re, _t), rd = Jo("disableVertical", () => {
  Qn(!1);
}), sd = Jo("enableVertical", () => {
  Qn(!0);
}), od = nd("mount", () => {
  Re() || (Qn(
    z() && U(gt) || !1,
    !0
  ), Re.set(!0));
});
function Qn(e, t) {
  (e !== Yt() || t) && (b(_t, { allow_vertical_swipe: e }), H(gt, e), Yt.set(e));
}
function ad() {
  Re.set(!1);
}
const Qf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  _isMounted: Re,
  _isVerticalEnabled: Yt,
  disableVertical: rd,
  enableVertical: sd,
  isMounted: Xp,
  isSupported: ed,
  isVerticalEnabled: td,
  mount: od,
  unmount: ad
}, Symbol.toStringTag, { value: "Module" })), Jf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  accentTextColor: al,
  backgroundColor: lt,
  bindCssVars: Ol,
  bottomBarBgColor: ko,
  buttonColor: Tn,
  buttonTextColor: So,
  destructiveTextColor: il,
  headerBackgroundColor: ul,
  hintColor: cl,
  isCssVarsBound: sl,
  isDark: ll,
  isMounted: Vl,
  isMounting: Ll,
  linkColor: pl,
  mount: Bo,
  mountError: Dl,
  mountPromise: jo,
  secondaryBackgroundColor: Qe,
  sectionBackgroundColor: dl,
  sectionHeaderTextColor: fl,
  sectionSeparatorColor: ml,
  state: ol,
  subtitleTextColor: hl,
  textColor: bl,
  unmount: Fl
}, Symbol.toStringTag, { value: "Module" })), me = "viewport", Jn = "fullscreen_changed", Kn = "safe_area_changed", Zn = "content_safe_area_changed", Yn = "viewport_changed", Ko = ce(me), qr = { left: 0, top: 0, bottom: 0, right: 0 };
function Bt(e) {
  return Math.max(e, 0);
}
const [jt, Zo] = /* @__PURE__ */ y({
  contentSafeAreaInsets: qr,
  height: 0,
  isExpanded: !1,
  isFullscreen: !1,
  safeAreaInsets: qr,
  stableHeight: 0,
  width: 0
});
function he(e) {
  return /* @__PURE__ */ h(() => Zo()[e]);
}
const Xn = he("height"), er = he("stableHeight"), Yo = he("width"), id = he("isExpanded"), ud = /* @__PURE__ */ h(() => Xn() === er());
function xe(e) {
  const { height: t, stableHeight: n, width: r } = e;
  jt.set({
    ...jt(),
    ...Pn({
      ...e,
      height: t ? Bt(t) : void 0,
      width: r ? Bt(r) : void 0,
      stableHeight: n ? Bt(n) : void 0
    })
  }), H(me, jt());
}
function cd() {
  return U(me);
}
function vt(e) {
  return /* @__PURE__ */ h(() => tr()[e]);
}
const tr = he("contentSafeAreaInsets"), Xo = vt("bottom"), ea = vt("left"), ta = vt("right"), na = vt("top");
function wt(e) {
  return /* @__PURE__ */ h(() => nr()[e]);
}
const nr = he("safeAreaInsets"), ra = wt("bottom"), sa = wt("left"), oa = wt("right"), aa = wt("top"), ia = "web_app_request_safe_area", ua = R(me, ia), Mr = ua(
  "requestContentSafeAreaInsets",
  (e) => E("web_app_request_content_safe_area", Zn, e)
);
function ld(e) {
  return E("web_app_request_viewport", Yn, e);
}
const Tr = ua(
  "requestSafeAreaInsets",
  (e) => E(ia, Kn, e)
), ca = (e) => {
  const { height: t } = e;
  xe({
    isExpanded: e.is_expanded,
    height: t,
    width: e.width,
    stableHeight: e.is_state_stable ? t : void 0
  });
}, la = (e) => {
  xe({ isFullscreen: e.is_fullscreen });
}, pa = (e) => {
  xe({ safeAreaInsets: e });
}, da = (e) => {
  xe({ contentSafeAreaInsets: e });
}, [
  pd,
  dd,
  fd,
  md
] = /* @__PURE__ */ Ve(
  me,
  (e) => {
    const t = z() && cd();
    return t ? j.resolve(t) : j.fn(async (n) => {
      const r = await j.all([
        Tr.isAvailable() ? Tr(n) : nr(),
        Mr.isAvailable() ? Mr(n) : tr()
      ]), s = dn(), o = {
        contentSafeAreaInsets: r[1],
        isFullscreen: !!s.tgWebAppFullscreen,
        safeAreaInsets: r[0]
      };
      if (["macos", "tdesktop", "unigram", "webk", "weba", "web"].includes(s.tgWebAppPlatform)) {
        const a = window;
        return {
          ...o,
          height: a.innerHeight,
          isExpanded: !0,
          stableHeight: a.innerHeight,
          width: a.innerWidth
        };
      }
      return ld(n).then((a) => ({
        ...o,
        height: a.height,
        isExpanded: a.is_expanded,
        stableHeight: a.is_state_stable ? a.height : 0,
        width: a.width
      }));
    }, e);
  },
  (e) => {
    C(Yn, ca), C(Jn, la), C(Kn, pa), C(Zn, da), xe(e);
  }
), hd = Ko("mount", pd), [, fa, bd] = dd, [, yd] = fd, [rr, _d] = md;
function gd() {
  Ce(fa), M(Yn, ca), M(Jn, la), M(Kn, pa), M(Zn, da), rr.set(!1);
}
const vd = at(me, rr), [qt, wd] = /* @__PURE__ */ y(!1), Sd = vd(
  "bindCssVars",
  (e) => {
    if (qt())
      throw new fn();
    e || (e = (n) => `--tg-viewport-${ys(n)}`);
    const t = [
      ["height", Xn],
      ["stableHeight", er],
      ["width", Yo],
      ["safeAreaInsetTop", aa],
      ["safeAreaInsetBottom", ra],
      ["safeAreaInsetLeft", sa],
      ["safeAreaInsetRight", oa],
      ["contentSafeAreaInsetTop", na],
      ["contentSafeAreaInsetBottom", Xo],
      ["contentSafeAreaInsetLeft", ea],
      ["contentSafeAreaInsetRight", ta]
    ].reduce((n, [r, s]) => {
      const o = e(r);
      if (o) {
        const a = () => {
          On(o, `${s()}px`);
        };
        n.push([a, s.sub(a), o]);
      }
      return n;
    }, []);
    return t.forEach((n) => {
      n[0]();
    }), qt.set(!0), () => {
      t.forEach((n) => {
        n[1](), Ln(n[2]);
      }), qt.set(!1);
    };
  }
), kd = Ko("expand", () => {
  b("web_app_expand");
}), ma = "web_app_request_fullscreen", Ed = ue(me, rr, ma), ha = he("isFullscreen"), [
  Ad,
  Cd
] = /* @__PURE__ */ y(), [
  xd,
  Pd
] = /* @__PURE__ */ y();
function ba(e, t) {
  return Ed(
    e,
    T(
      (n) => E(
        t ? ma : "web_app_exit_fullscreen",
        [Jn, "fullscreen_failed"],
        n
      ).then((r) => {
        if ("error" in r && r.error !== "ALREADY_FULLSCREEN")
          throw new Ri(r.error);
        const s = "is_fullscreen" in r ? r.is_fullscreen : !0;
        s !== ha() && xe({ isFullscreen: s });
      }),
      "Fullscreen mode change is already being requested",
      {
        promise: Ad,
        error: xd
      }
    )[0]
  );
}
const Bd = ba("requestFullscreen", !0), jd = ba("exitFullscreen"), Kf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  bindCssVars: Sd,
  changeFullscreenError: Pd,
  changeFullscreenPromise: Cd,
  contentSafeAreaInsetBottom: Xo,
  contentSafeAreaInsetLeft: ea,
  contentSafeAreaInsetRight: ta,
  contentSafeAreaInsetTop: na,
  contentSafeAreaInsets: tr,
  exitFullscreen: jd,
  expand: kd,
  height: Xn,
  isCssVarsBound: wd,
  isExpanded: id,
  isFullscreen: ha,
  isMounted: _d,
  isMounting: bd,
  isStable: ud,
  mount: hd,
  mountError: yd,
  mountPromise: fa,
  requestFullscreen: Bd,
  safeAreaInsetBottom: ra,
  safeAreaInsetLeft: sa,
  safeAreaInsetRight: oa,
  safeAreaInsetTop: aa,
  safeAreaInsets: nr,
  stableHeight: er,
  state: Zo,
  unmount: gd,
  width: Yo
}, Symbol.toStringTag, { value: "Module" })), ya = "web_app_request_emoji_status_access", [
  qd,
  Md,
  Td
] = T((e) => E(ya, "emoji_status_access_requested", e).then((t) => t.status), "Emoji status access request is already in progress"), Zf = /* @__PURE__ */ A(
  "requestEmojiStatusAccess",
  qd,
  { isSupported: ya }
), [, Yf, Xf] = Md, [, em] = Td, _a = "web_app_set_emoji_status", [
  $d,
  Id,
  Od
] = T(
  (e, t) => E(_a, ["emoji_status_set", "emoji_status_failed"], {
    params: {
      custom_emoji_id: e,
      duration: (t || {}).duration
    },
    ...t
  }).then((n) => {
    if (n && "error" in n)
      throw new Di(n.error);
  }),
  "Emoji status set request is currently in progress"
), tm = /* @__PURE__ */ A("setEmojiStatus", $d, {
  isSupported: _a
}), [, nm, rm] = Id, [, sm] = Od, Ld = { isSupported: "web_app_add_to_home_screen" }, ga = "home_screen_failed", om = /* @__PURE__ */ A(
  "onAddToHomeScreenFailed",
  (e, t) => C(ga, e, t),
  { isSupported: "web_app_add_to_home_screen" }
), am = /* @__PURE__ */ A(
  "offAddToHomeScreenFailed",
  (e) => {
    M(ga, e);
  },
  Ld
), va = { isSupported: "web_app_add_to_home_screen" }, wa = "home_screen_added", im = /* @__PURE__ */ A(
  "onAddedToHomeScreen",
  (e, t) => C(wa, e, t),
  va
), um = /* @__PURE__ */ A(
  "offAddedToHomeScreen",
  (e) => {
    M(wa, e);
  },
  va
), $r = "web_app_add_to_home_screen", cm = /* @__PURE__ */ A(
  "addToHomeScreen",
  () => {
    b($r);
  },
  { isSupported: $r }
), Sa = "web_app_check_home_screen", [
  Dd,
  Rd,
  Vd
] = T((e) => E(Sa, "home_screen_checked", e).then((t) => t.status || "unknown"), "Check home screen status request is currently in progress"), lm = /* @__PURE__ */ A("checkHomeScreenStatus", Dd, {
  isSupported: Sa
}), [, pm, dm] = Rd, [, fm] = Vd, Fd = ce(), mm = Fd(
  "openLink",
  (e, t) => {
    if (typeof e == "string")
      try {
        e = new URL(e);
      } catch (n) {
        throw new J(`"${e.toString()}" is invalid URL`, n);
      }
    t || (t = {}), b("web_app_open_link", {
      url: e.toString(),
      try_browser: t.tryBrowser,
      try_instant_view: t.tryInstantView
    });
  }
), Ir = "web_app_open_tg_link", Wd = ce(), Nd = Wd(
  "openTelegramLink",
  (e) => {
    const t = e.toString();
    if (!t.match(/^https:\/\/t.me\/.+/))
      throw new J(`"${t}" is invalid URL`);
    if (!pe(Ir, re())) {
      window.location.href = t;
      return;
    }
    e = new URL(e), b(Ir, { path_full: e.pathname + e.search });
  }
), Hd = ce(), hm = Hd(
  "shareURL",
  (e, t) => {
    Nd(
      "https://t.me/share/url?" + new URLSearchParams({ url: e, text: t || "" }).toString().replace(/\+/g, "%20")
    );
  }
);
function Ud(e, t) {
  return new j({ abortSignal: t, timeout: e }).catch(() => {
  });
}
const ka = "web_app_request_phone", [
  zd,
  Gd,
  Qd
] = T((e) => E(ka, "phone_requested", e).then((t) => t.status), "Phone access request is currently in progress"), Jd = /* @__PURE__ */ A("requestPhoneAccess", zd, {
  isSupported: ka
}), [, bm, ym] = Gd, [, _m] = Qd;
async function Or(e) {
  const t = await Ae("getRequestedContact", {}, {
    ...e,
    timeout: (e || {}).timeout || 5e3
  });
  return ut(
    /* @__PURE__ */ Te(
      // todo: Union is unnecessary here, but we use it to comply TypeScript checker.
      /* @__PURE__ */ Qs([/* @__PURE__ */ W(), /* @__PURE__ */ Us(URLSearchParams)]),
      yo(
        /* @__PURE__ */ zt({
          contact: /* @__PURE__ */ Te(
            /* @__PURE__ */ W(),
            bo(),
            /* @__PURE__ */ zt({
              user_id: /* @__PURE__ */ kn(),
              phone_number: /* @__PURE__ */ W(),
              first_name: /* @__PURE__ */ W(),
              last_name: /* @__PURE__ */ zs(/* @__PURE__ */ W())
            })
          ),
          auth_date: /* @__PURE__ */ Te(
            /* @__PURE__ */ W(),
            /* @__PURE__ */ wn((n) => new Date(Number(n) * 1e3)),
            /* @__PURE__ */ Sn()
          ),
          hash: /* @__PURE__ */ W()
        })
      )
    ),
    t
  );
}
const [
  Kd,
  Zd,
  Yd
] = T(
  (e) => new j(async (t, n, r) => {
    try {
      return t(await Or(r));
    } catch (o) {
      if (o instanceof Ut)
        throw o;
    }
    if (await Jd(r) !== "sent")
      throw new Es("User denied access");
    let s = 50;
    for (; !r.isAborted(); ) {
      try {
        return t(await Or(r));
      } catch (o) {
        if (o instanceof Ut)
          throw o;
      }
      await Ud(s), s += 50;
    }
  }, e),
  "Contact is already being requested"
), gm = /* @__PURE__ */ A("requestContact", Kd, {
  isSupported: "web_app_request_phone"
}), [, vm, wm] = Zd, [, Sm] = Yd, Ea = "web_app_request_write_access", [
  Xd,
  ef,
  tf
] = T(
  (e) => E(Ea, "write_access_requested", e).then((t) => t.status),
  "Write access request is currently in progress"
), km = /* @__PURE__ */ A("requestWriteAccess", Xd, {
  isSupported: Ea
}), [, Em, Am] = ef, [, Cm] = tf;
function nf(e) {
  const t = document.createElement("textarea");
  t.value = e, t.style.top = "0", t.style.left = "0", t.style.position = "fixed", document.body.appendChild(t), t.focus(), t.select();
  try {
    document.execCommand("copy");
  } finally {
    document.body.removeChild(t);
  }
}
async function xm(e) {
  try {
    const { clipboard: t } = navigator;
    if (t)
      return await t.writeText(e);
  } catch {
  }
  nf(e);
}
const Lr = "web_app_request_file_download", Pm = /* @__PURE__ */ A(
  "downloadFile",
  (e, t, n) => E(
    Lr,
    "file_download_requested",
    { ...n, params: { url: e, file_name: t } }
  ).then((r) => {
    if (r.status !== "downloading")
      throw new Es("User denied the action");
  }),
  { isSupported: Lr }
), Bm = /* @__PURE__ */ A(
  "getCurrentTime",
  (e) => Ae("getCurrentTime", {}, e).then((t) => ut(
    /* @__PURE__ */ Te(/* @__PURE__ */ kn(), /* @__PURE__ */ Ws(), /* @__PURE__ */ wn((n) => new Date(n * 1e3)), /* @__PURE__ */ Sn()),
    t
  )),
  { isSupported: "web_app_invoke_custom_method" }
), Dr = "web_app_read_text_from_clipboard", jm = /* @__PURE__ */ A(
  "readTextFromClipboard",
  (e) => {
    const t = Ss();
    return E(Dr, "clipboard_text_received", {
      ...e,
      params: { req_id: t },
      capture: bs(t)
    }).then(({ data: n = null }) => n);
  },
  { isSupported: Dr }
), qm = /* @__PURE__ */ A(
  "sendData",
  (e) => {
    const { size: t } = new Blob([e]);
    if (!t || t > 4096)
      throw new J(t ? "Maximum size of data to send is 4096 bytes" : "Attempted to send empty data");
    b("web_app_data_send", { data: e });
  },
  {
    isSupported() {
      return dn().tgWebAppBotInline ? void 0 : "The application must be launched in the inline mode";
    }
  }
), Rr = "web_app_send_prepared_message", Mm = /* @__PURE__ */ A(
  "shareMessage",
  (e, t) => E(Rr, ["prepared_message_failed", "prepared_message_sent"], {
    ...t,
    params: { id: e }
  }).then((n) => {
    if (n && "error" in n)
      throw new Vi(n.error);
  }),
  { isSupported: Rr }
), Vr = "web_app_share_to_story", Tm = /* @__PURE__ */ A(
  "shareStory",
  (e, t) => {
    t || (t = {}), b(Vr, {
      text: t.text,
      media_url: e,
      widget_link: t.widgetLink
    });
  },
  { isSupported: Vr }
), rf = "web_app_switch_inline_query", $m = /* @__PURE__ */ A(
  "switchInlineQuery",
  (e, t) => {
    b(rf, {
      query: e,
      chat_types: t || []
    });
  }
);
function Im(e) {
  try {
    return [!0, e()];
  } catch (t) {
    return [!1, t];
  }
}
function Om(e) {
  Ii(e);
  const [t, n] = pn(
    C("reload_iframe", () => {
      Vt(!1, "Received a request to reload the page"), b("iframe_will_reload"), window.location.reload();
    })
  ), { acceptCustomStyles: r = !0 } = e || {};
  if (r) {
    const s = document.createElement("style");
    s.id = "telegram-custom-styles", document.head.appendChild(s), t(
      C("set_custom_style", (o) => {
        s.innerHTML = o;
      }),
      () => {
        document.head.removeChild(s);
      }
    );
  }
  return b("iframe_ready", { reload_supported: !0 }), Vt(!1, "The package was initialized"), n;
}
export {
  Es as AccessDeniedError,
  fn as CSSVarsBoundError,
  Li as ConcurrentCallError,
  Ri as FullscreenFailedError,
  Pe as FunctionUnavailableError,
  J as InvalidArgumentsError,
  wf as InvalidEnvError,
  _i as InvalidLaunchParamsError,
  vi as InvokeCustomMethodError,
  yi as LaunchParamsRetrieveError,
  hi as MethodParameterUnsupportedError,
  mi as MethodUnsupportedError,
  ks as NotAvailableError,
  Di as SetEmojiStatusError,
  Vi as ShareMessageError,
  gi as UnknownEnvError,
  cm as addToHomeScreen,
  gu as authenticateBiometry,
  jf as backButton,
  Xl as bindMiniAppCssVars,
  Ol as bindThemeParamsCssVars,
  Sd as bindViewportCssVars,
  Mf as biometry,
  wu as biometryAuthError,
  Is as biometryAuthPromise,
  fu as biometryMountError,
  su as biometryState,
  Pd as changeFullscreenError,
  Cd as changeFullscreenPromise,
  lm as checkHomeScreenStatus,
  fm as checkHomeScreenStatusError,
  pm as checkHomeScreenStatusPromise,
  ep as closeMiniApp,
  vp as closeQrScanner,
  Tf as closingBehavior,
  $f as cloudStorage,
  Ii as configure,
  xm as copyTextToClipboard,
  xi as createPostEvent,
  Ss as createRequestId,
  Ks as deleteCloudStorageItem,
  Tu as disableClosingConfirmation,
  rd as disableVerticalSwipes,
  Pm as downloadFile,
  lr as emitEvent,
  $u as enableClosingConfirmation,
  sd as enableVerticalSwipes,
  jd as exitFullscreen,
  kd as expandViewport,
  Wu as getCloudStorageItem,
  Zs as getCloudStorageKeys,
  Bm as getCurrentTime,
  If as hapticFeedback,
  zu as hapticFeedbackImpactOccurred,
  Gu as hapticFeedbackNotificationOccurred,
  Qu as hapticFeedbackSelectionChanged,
  Hi as hideBackButton,
  Gp as hideSettingsButton,
  Bs as ignoreCanceled,
  Om as init,
  Of as initData,
  Xs as initDataAuthDate,
  eo as initDataCanSendAfter,
  Ku as initDataCanSendAfterDate,
  Zu as initDataChat,
  Xu as initDataChatInstance,
  Yu as initDataChatType,
  ec as initDataHash,
  tc as initDataQueryId,
  rc as initDataRaw,
  sc as initDataReceiver,
  ac as initDataStartParam,
  Ju as initDataState,
  ic as initDataUser,
  Lf as invoice,
  Ae as invokeCustomMethod,
  xf as isAccessDeniedError,
  vu as isAuthenticatingBiometry,
  Wi as isBackButtonMounted,
  Ni as isBackButtonSupported,
  Fi as isBackButtonVisible,
  ou as isBiometryAvailable,
  hu as isBiometryMounted,
  du as isBiometryMounting,
  iu as isBiometrySupported,
  gf as isCSSVarsBoundError,
  dm as isCheckingHomeScreenStatus,
  Ls as isClosingBehaviorMounted,
  qu as isClosingConfirmationEnabled,
  Vu as isCloudStorageSupported,
  wo as isColorDark,
  Af as isConcurrentCallError,
  ha as isFullscreen,
  Pf as isFullscreenFailedError,
  kf as isFunctionNotAvailableError,
  Uu as isHapticFeedbackSupported,
  Ef as isInvalidArguments,
  Sf as isInvalidEnvError,
  ff as isInvalidLaunchParamsError,
  bc as isInvoiceOpened,
  cc as isInvoiceSupported,
  hf as isInvokeCustomMethodError,
  df as isLaunchParamsRetrieveError,
  vc as isLocationManagerAccessGranted,
  wc as isLocationManagerAccessRequested,
  gc as isLocationManagerAvailable,
  Mc as isLocationManagerMounted,
  Bc as isLocationManagerMounting,
  _c as isLocationManagerSupported,
  gl as isMainButtonEnabled,
  vl as isMainButtonLoaderVisible,
  Eo as isMainButtonMounted,
  wl as isMainButtonVisible,
  pf as isMethodMethodParameterUnsupportedError,
  lf as isMethodUnsupportedError,
  Gl as isMiniAppActive,
  Ul as isMiniAppCssVarsBound,
  zl as isMiniAppDark,
  op as isMiniAppMounted,
  np as isMiniAppMounting,
  Ql as isMiniAppSupported,
  vf as isNotAvailableError,
  bp as isPopupOpened,
  lp as isPopupSupported,
  xp as isQrScannerOpened,
  wp as isQrScannerSupported,
  ve as isRGB,
  Yc as isRGBShort,
  xu as isRequestingBiometryAccess,
  wm as isRequestingContact,
  Xf as isRequestingEmojiStatusAccess,
  Lc as isRequestingLocation,
  ym as isRequestingPhoneAccess,
  Am as isRequestingWriteAccess,
  _r as isSSR,
  qp as isSecondaryButtonEnabled,
  Mp as isSecondaryButtonLoaderVisible,
  Uo as isSecondaryButtonMounted,
  Dp as isSecondaryButtonSupported,
  Tp as isSecondaryButtonVisible,
  Cf as isSetEmojiStatusError,
  rm as isSettingEmojiStatus,
  Up as isSettingsButtonMounted,
  zp as isSettingsButtonSupported,
  Hp as isSettingsButtonVisible,
  Bf as isShareMessageError,
  Xp as isSwipeBehaviorMounted,
  ed as isSwipeBehaviorSupported,
  hr as isTMA,
  sl as isThemeParamsCssVarsBound,
  ll as isThemeParamsDark,
  Vl as isThemeParamsMounted,
  Ll as isThemeParamsMounting,
  mf as isUnknownEnvError,
  td as isVerticalSwipesEnabled,
  wd as isViewportCssVarsBound,
  id as isViewportExpanded,
  _d as isViewportMounted,
  bd as isViewportMounting,
  ud as isViewportStable,
  Df as locationManager,
  jc as locationManagerMountError,
  Pc as locationManagerMountPromise,
  Wf as mainButton,
  yl as mainButtonBackgroundColor,
  _l as mainButtonHasShineEffect,
  $n as mainButtonState,
  Sl as mainButtonText,
  kl as mainButtonTextColor,
  Nf as miniApp,
  Wl as miniAppBackgroundColor,
  Dn as miniAppBackgroundColorRGB,
  Nl as miniAppBottomBarColor,
  Rn as miniAppBottomBarColorRGB,
  Hl as miniAppHeaderColor,
  Mo as miniAppHeaderColorRGB,
  rp as miniAppMountError,
  Do as miniAppMountPromise,
  ap as miniAppReady,
  To as miniAppState,
  bf as mockTelegramEnv,
  Ui as mountBackButton,
  pu as mountBiometry,
  $s as mountBiometryPromise,
  Iu as mountClosingBehavior,
  xc as mountLocationManager,
  Cl as mountMainButton,
  tp as mountMiniApp,
  Rp as mountSecondaryButton,
  Qp as mountSettingsButton,
  od as mountSwipeBehavior,
  Bo as mountThemeParams,
  hd as mountViewport,
  M as off,
  am as offAddToHomeScreenFailed,
  um as offAddedToHomeScreen,
  Gi as offBackButtonClick,
  Pl as offMainButtonClick,
  Fp as offSecondaryButtonClick,
  Kp as offSettingsButtonClick,
  C as on,
  om as onAddToHomeScreenFailed,
  im as onAddedToHomeScreen,
  zi as onBackButtonClick,
  xl as onMainButtonClick,
  Vp as onSecondaryButtonClick,
  Jp as onSettingsButtonClick,
  Su as openBiometrySettings,
  mc as openInvoice,
  yc as openInvoiceError,
  hc as openInvoicePromise,
  mm as openLink,
  Rc as openLocationManagerSettings,
  mp as openPopup,
  yp as openPopupError,
  hp as openPopupPromise,
  Cp as openQrScanner,
  Pp as openQrScannerError,
  Ho as openQrScannerPromise,
  Nd as openTelegramLink,
  Rf as parseInitDataQuery,
  Vf as parseLaunchParamsQuery,
  Hf as popup,
  b as postEvent,
  Uf as qrScanner,
  jm as readTextFromClipboard,
  E as request,
  au as requestBiometry,
  Cu as requestBiometryAccess,
  Pu as requestBiometryAccessError,
  Os as requestBiometryAccessPromise,
  gm as requestContact,
  Sm as requestContactError,
  vm as requestContactPromise,
  Mr as requestContentSafeAreaInsets,
  Zf as requestEmojiStatusAccess,
  em as requestEmojiStatusAccessError,
  Yf as requestEmojiStatusAccessPromise,
  Bd as requestFullscreen,
  Oc as requestLocation,
  Dc as requestLocationError,
  ro as requestLocationPromise,
  Jd as requestPhoneAccess,
  _m as requestPhoneAccessError,
  bm as requestPhoneAccessPromise,
  Tr as requestSafeAreaInsets,
  ld as requestViewport,
  km as requestWriteAccess,
  Cm as requestWriteAccessError,
  Em as requestWriteAccessPromise,
  oc as restoreInitData,
  ln as retrieveLaunchParams,
  Ai as retrieveRawInitData,
  hs as retrieveRawLaunchParams,
  Im as safeCall,
  zf as secondaryButton,
  Bp as secondaryButtonBackgroundColor,
  jp as secondaryButtonHasShineEffect,
  $p as secondaryButtonPosition,
  Hn as secondaryButtonState,
  Ip as secondaryButtonText,
  Op as secondaryButtonTextColor,
  qm as sendData,
  rl as serializeInitDataQuery,
  Ff as serializeLaunchParamsQuery,
  vo as serializeToQuery,
  Nu as setCloudStorageItem,
  _f as setDebug,
  tm as setEmojiStatus,
  sm as setEmojiStatusError,
  nm as setEmojiStatusPromise,
  Bl as setMainButtonParams,
  Ro as setMiniAppBackgroundColor,
  Vo as setMiniAppBottomBarColor,
  Fo as setMiniAppHeaderColor,
  Wp as setSecondaryButtonParams,
  Gf as settingsButton,
  Mm as shareMessage,
  Tm as shareStory,
  hm as shareURL,
  Qi as showBackButton,
  Zp as showSettingsButton,
  pe as supports,
  Qf as swipeBehavior,
  $m as switchInlineQuery,
  Ei as targetOrigin,
  Jf as themeParams,
  al as themeParamsAccentTextColor,
  lt as themeParamsBackgroundColor,
  ko as themeParamsBottomBarBgColor,
  Tn as themeParamsButtonColor,
  So as themeParamsButtonTextColor,
  il as themeParamsDestructiveTextColor,
  ul as themeParamsHeaderBackgroundColor,
  cl as themeParamsHintColor,
  pl as themeParamsLinkColor,
  Dl as themeParamsMountError,
  jo as themeParamsMountPromise,
  Qe as themeParamsSecondaryBackgroundColor,
  dl as themeParamsSectionBackgroundColor,
  fl as themeParamsSectionHeaderTextColor,
  ml as themeParamsSectionSeparatorColor,
  ol as themeParamsState,
  hl as themeParamsSubtitleTextColor,
  bl as themeParamsTextColor,
  Xc as toRGB,
  yo as transformQueryUsing,
  Ji as unmountBackButton,
  Bu as unmountBiometry,
  Ou as unmountClosingBehavior,
  Vc as unmountLocationManager,
  jl as unmountMainButton,
  ip as unmountMiniApp,
  Np as unmountSecondaryButton,
  Yp as unmountSettingsButton,
  ad as unmountSwipeBehavior,
  Fl as unmountThemeParams,
  gd as unmountViewport,
  ju as updateBiometryToken,
  of as useSignal,
  Kf as viewport,
  Xo as viewportContentSafeAreaInsetBottom,
  ea as viewportContentSafeAreaInsetLeft,
  ta as viewportContentSafeAreaInsetRight,
  na as viewportContentSafeAreaInsetTop,
  tr as viewportContentSafeAreaInsets,
  Xn as viewportHeight,
  yd as viewportMountError,
  fa as viewportMountPromise,
  ra as viewportSafeAreaInsetBottom,
  sa as viewportSafeAreaInsetLeft,
  oa as viewportSafeAreaInsetRight,
  aa as viewportSafeAreaInsetTop,
  nr as viewportSafeAreaInsets,
  er as viewportStableHeight,
  Zo as viewportState,
  Yo as viewportWidth,
  A as wrapSafe
};
